<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="From raw ChIP-seq data to model input.">

<title>BPNet - 01 Prepare Input</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">BPNet</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">Home</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./00_introduction.html">Introduction</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./01_prepare_input.html" aria-current="page">Data Processing</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./02_APIs.html">APIs</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./03_training.html">Training</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./04_architecture_comparison.html">Architectures</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./05_interpretability.qmd">Interpretability</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./06_mutagenesis.qmd">Mutagenesis</a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#libraries" id="toc-libraries" class="nav-link active" data-scroll-target="#libraries"><span class="toc-section-number">1</span>  Libraries</a></li>
  <li><a href="#configs" id="toc-configs" class="nav-link" data-scroll-target="#configs"><span class="toc-section-number">2</span>  Configs</a>
  <ul class="collapse">
  <li><a href="#output-directory" id="toc-output-directory" class="nav-link" data-scroll-target="#output-directory"><span class="toc-section-number">2.1</span>  Output Directory</a></li>
  <li><a href="#peak-width" id="toc-peak-width" class="nav-link" data-scroll-target="#peak-width"><span class="toc-section-number">2.2</span>  Peak Width</a></li>
  <li><a href="#chromosome-traintunetest-split" id="toc-chromosome-traintunetest-split" class="nav-link" data-scroll-target="#chromosome-traintunetest-split"><span class="toc-section-number">2.3</span>  Chromosome Train/Tune/Test split</a></li>
  <li><a href="#colors" id="toc-colors" class="nav-link" data-scroll-target="#colors"><span class="toc-section-number">2.4</span>  Colors</a></li>
  </ul></li>
  <li><a href="#function-for-one-hot-encoding" id="toc-function-for-one-hot-encoding" class="nav-link" data-scroll-target="#function-for-one-hot-encoding"><span class="toc-section-number">3</span>  Function for One-Hot Encoding</a></li>
  <li><a href="#read-stats" id="toc-read-stats" class="nav-link" data-scroll-target="#read-stats"><span class="toc-section-number">4</span>  Read Stats</a></li>
  <li><a href="#load-the-peak-data" id="toc-load-the-peak-data" class="nav-link" data-scroll-target="#load-the-peak-data"><span class="toc-section-number">5</span>  Load the Peak Data</a>
  <ul class="collapse">
  <li><a href="#distribution-of-the-macs2-scores" id="toc-distribution-of-the-macs2-scores" class="nav-link" data-scroll-target="#distribution-of-the-macs2-scores"><span class="toc-section-number">5.1</span>  Distribution of the MACS2 Scores</a></li>
  <li><a href="#number-of-peaks-per-tf" id="toc-number-of-peaks-per-tf" class="nav-link" data-scroll-target="#number-of-peaks-per-tf"><span class="toc-section-number">5.2</span>  Number of Peaks per TF</a></li>
  </ul></li>
  <li><a href="#translate-peak-sequences-to-one-hot-encoding" id="toc-translate-peak-sequences-to-one-hot-encoding" class="nav-link" data-scroll-target="#translate-peak-sequences-to-one-hot-encoding"><span class="toc-section-number">6</span>  Translate Peak Sequences to One-Hot Encoding</a></li>
  <li><a href="#extract-the-tf-counts" id="toc-extract-the-tf-counts" class="nav-link" data-scroll-target="#extract-the-tf-counts"><span class="toc-section-number">7</span>  Extract the TF Counts</a>
  <ul class="collapse">
  <li><a href="#examples-highest-scoring-peaks" id="toc-examples-highest-scoring-peaks" class="nav-link" data-scroll-target="#examples-highest-scoring-peaks"><span class="toc-section-number">7.1</span>  Examples: Highest Scoring Peaks</a></li>
  <li><a href="#examples-random-peaks" id="toc-examples-random-peaks" class="nav-link" data-scroll-target="#examples-random-peaks"><span class="toc-section-number">7.2</span>  Examples: Random Peaks</a></li>
  </ul></li>
  <li><a href="#extract-the-control-patchcap-counts" id="toc-extract-the-control-patchcap-counts" class="nav-link" data-scroll-target="#extract-the-control-patchcap-counts"><span class="toc-section-number">8</span>  Extract the Control (Patchcap) Counts</a>
  <ul class="collapse">
  <li><a href="#examples-highest-scoring-peaks-with-bias" id="toc-examples-highest-scoring-peaks-with-bias" class="nav-link" data-scroll-target="#examples-highest-scoring-peaks-with-bias"><span class="toc-section-number">8.1</span>  Examples: Highest Scoring Peaks with Bias</a></li>
  <li><a href="#examples-random-peaks-with-bias" id="toc-examples-random-peaks-with-bias" class="nav-link" data-scroll-target="#examples-random-peaks-with-bias"><span class="toc-section-number">8.2</span>  Examples: Random Peaks with Bias</a></li>
  </ul></li>
  <li><a href="#figure-1e" id="toc-figure-1e" class="nav-link" data-scroll-target="#figure-1e"><span class="toc-section-number">9</span>  Figure 1e</a>
  <ul class="collapse">
  <li><a href="#from-raw-data" id="toc-from-raw-data" class="nav-link" data-scroll-target="#from-raw-data"><span class="toc-section-number">9.1</span>  From Raw Data</a></li>
  <li><a href="#from-processed-data" id="toc-from-processed-data" class="nav-link" data-scroll-target="#from-processed-data"><span class="toc-section-number">9.2</span>  From Processed Data</a></li>
  </ul></li>
  <li><a href="#oct4-sox2-motif" id="toc-oct4-sox2-motif" class="nav-link" data-scroll-target="#oct4-sox2-motif"><span class="toc-section-number">10</span>  Oct4-Sox2 Motif</a></li>
  <li><a href="#appendix" id="toc-appendix" class="nav-link" data-scroll-target="#appendix"><span class="toc-section-number">11</span>  Appendix</a>
  <ul class="collapse">
  <li><a href="#save-image" id="toc-save-image" class="nav-link" data-scroll-target="#save-image"><span class="toc-section-number">11.1</span>  Save Image</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info"><span class="toc-section-number">11.2</span>  Session Info</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">01 Prepare Input</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source" data-quarto-source-url="repo"><i class="bi"></i> Code</button></div></div>
</div>

<div>
  <div class="description">
    From raw ChIP-seq data to model input.
  </div>
</div>


<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<section id="libraries" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Libraries</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">RETICULATE_PYTHON =</span> <span class="st">"/home/philipp/miniconda3/envs/r-reticulate/bin/python3"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(reticulate)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(rtracklayer)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(GenomicAlignments)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(BRGenomics)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(BSgenome.Mmusculus.UCSC.mm10)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(tidyverse)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(furrr)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(motifmatchr)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(JASPAR2020)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(TFBSTools)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Setup Multiprocessing</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize=</span><span class="dv">5000</span><span class="sc">*</span><span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>future<span class="sc">::</span><span class="fu">plan</span>(future<span class="sc">::</span>multisession, <span class="at">workers =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="configs" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Configs</h1>
<section id="output-directory" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="output-directory"><span class="header-section-number">2.1</span> Output Directory</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="st">"/home/philipp/BPNet/input/"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(output_dir)){<span class="fu">dir.create</span>(output_dir)}</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>TFs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Sox2"</span>, <span class="st">"Oct4"</span>, <span class="st">"Klf4"</span>, <span class="st">"Nanog"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>purrr<span class="sc">::</span><span class="fu">walk</span>(<span class="fu">c</span>(TFs, <span class="st">"patchcap"</span>, <span class="st">"figures"</span>), <span class="sc">~</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="fu">paste0</span>(output_dir, .x))){<span class="fu">dir.create</span>(<span class="fu">paste0</span>(output_dir, .x))})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="peak-width" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="peak-width"><span class="header-section-number">2.2</span> Peak Width</h2>
<p>As in the BPNet paper, we are using a peak width of 1000 bp, meaning we consider 500 bp up- and downstream of the ChIP-seq peaks.</p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-3_cafe177374371d32ee912394e3d511c8">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>peak_width <span class="ot">&lt;-</span> <span class="dv">1000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="chromosome-traintunetest-split" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="chromosome-traintunetest-split"><span class="header-section-number">2.3</span> Chromosome Train/Tune/Test split</h2>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-4_7f4b0cf4060801d739815db0a1af7eb6">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>all_chroms <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"chr"</span>, <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">19</span>, <span class="st">"X"</span>, <span class="st">"Y"</span>))</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># we use the same train/tune/test split as in the BPnet paper</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>chrom_list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">"tune"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"chr2"</span>, <span class="st">"chr3"</span>, <span class="st">"chr4"</span>), <span class="co"># tune set (hyperparameter tuning): chromosomes 2, 3, 4</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"test"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr8"</span>, <span class="st">"chr9"</span>)) <span class="co"># test set (performance evaluation): chromosome 1, 8, 9</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>chrom_list<span class="sc">$</span>train <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(all_chroms, <span class="fu">c</span>(chrom_list<span class="sc">$</span>tune, chrom_list<span class="sc">$</span>test)) <span class="co"># train set: all other chromosomes</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>chrom_list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$tune
[1] "chr2" "chr3" "chr4"

$test
[1] "chr1" "chr8" "chr9"

$train
 [1] "chr5"  "chr6"  "chr7"  "chr10" "chr11" "chr12" "chr13" "chr14" "chr15"
[10] "chr16" "chr17" "chr18" "chr19" "chrX"  "chrY" </code></pre>
</div>
</div>
</section>
<section id="colors" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="colors"><span class="header-section-number">2.4</span> Colors</h2>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-5_9982b2d3e71bcea0ed823651b3fa533c">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Klf4"</span> <span class="ot">=</span> <span class="st">"#92C592"</span>,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>           <span class="st">"Nanog"</span> <span class="ot">=</span> <span class="st">"#FFE03F"</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>           <span class="st">"Oct4"</span> <span class="ot">=</span> <span class="st">"#CD5C5C"</span>,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>           <span class="st">"Sox2"</span> <span class="ot">=</span> <span class="st">"#849EEB"</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>           <span class="st">"patchcap"</span> <span class="ot">=</span> <span class="st">"#827F81"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="function-for-one-hot-encoding" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Function for One-Hot Encoding</h1>
<p>Traditionally DNA sequences are encoded using: A = [1, 0, 0, 0], C = [0, 1, 0, 0], G = [0, 0, 1, 0], T = [0, 0, 0, 1], N = [0, 0, 0, 0].</p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-6_1f5498c55e126be5ded9997dd1e22ca6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>one_hot <span class="ot">&lt;-</span> <span class="cf">function</span>(sequence) {</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  len <span class="ot">=</span> <span class="fu">nchar</span>(sequence)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  mtx <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="dv">0</span>, <span class="at">nrow=</span><span class="dv">4</span>, <span class="at">ncol=</span>len)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>len) {</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">substr</span>(sequence, i, i) <span class="sc">==</span> <span class="st">"A"</span>) mtx[<span class="dv">1</span>, i] <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span> <span class="cf">if</span> (<span class="fu">substr</span>(sequence, i, i) <span class="sc">==</span> <span class="st">"C"</span>) mtx[<span class="dv">2</span>, i] <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span> <span class="cf">if</span> (<span class="fu">substr</span>(sequence, i, i) <span class="sc">==</span> <span class="st">"G"</span>) mtx[<span class="dv">3</span>, i] <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span> <span class="cf">if</span> (<span class="fu">substr</span>(sequence, i, i) <span class="sc">==</span> <span class="st">"T"</span>) mtx[<span class="dv">4</span>, i] <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(mtx)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>compiler<span class="sc">::</span><span class="fu">enableJIT</span>(<span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>one_hot_c <span class="ot">&lt;-</span> compiler<span class="sc">::</span><span class="fu">cmpfun</span>(one_hot) <span class="co"># compiling to bytecode not machine code like numba</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="read-stats" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Read Stats</h1>
<p>Look at the reading stats per TF to check how comparable the different ChIP-Nexus experiments are.</p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-7_fff1cb96d98c8c6572609abe9b8f08c8">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>total_reads <span class="ot">&lt;-</span> TFs <span class="sc">%&gt;%</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">set_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map_dbl</span>(<span class="cf">function</span>(tf) {</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    Rsamtools<span class="sc">::</span><span class="fu">countBam</span>(<span class="fu">paste0</span>(<span class="st">"../data/chip-nexus/"</span>, tf, <span class="st">"/pool_filt.bam"</span>))<span class="sc">$</span>records</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  })</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">y=</span><span class="fu">names</span>(total_reads), <span class="at">x=</span>total_reads<span class="sc">/</span><span class="fl">1e6</span>), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">fill=</span><span class="st">"blue"</span>, <span class="at">alpha=</span><span class="fl">0.4</span>, <span class="at">width=</span><span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"N"</span>, <span class="at">y=</span><span class="st">"TF"</span>) <span class="sc">+</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Inverted geom defaults of fill and color/colour.
To change them back, use invert_geom_defaults().</code></pre>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="load-the-peak-data" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Load the Peak Data</h1>
<p>For each transcription factor (TF) we read in the peak information provided by the authors of the BPNet paper. The peaks were called using <code>MACS2</code> (v.2.1.1.20160309). The following non-default settings were used: <code>shift=-75</code>, <code>extsize=150</code>, meaning the 5’ ends of each read were extended 75 bp in both directions.</p>
<p>In particular <code>shift=-75</code> means the 5’ ends of the reads are moved -75 bp and <code>extsize=150</code> means the reads are extended to a fixed length of 150 bp in 5’ -&gt; 3’ direction.</p>
<details>
<summary>
Click to view sketch
</summary>
<img src="img/macs2_explanation.png" class="img-fluid">
</details>
<p>Check the <code>macs3 callpeak</code> documentation <a href="https://github.com/macs3-project/MACS/blob/master/docs/callpeak.md">here</a></p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-9_379904c07f5851aa0fc7a99cffd367e2">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>peak_infos <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(TFs, <span class="cf">function</span>(tf) {</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  narrowPeaks <span class="ot">&lt;-</span> </span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    rtracklayer<span class="sc">::</span><span class="fu">import</span>(<span class="fu">paste0</span>(<span class="st">"/home/philipp/AML_Final_Project/data/chip-nexus/"</span>, tf, <span class="st">"/idr-optimal-set.narrowPeak"</span>), </span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>                           <span class="at">format=</span><span class="st">"narrowPeak"</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># read in the peak summits and extend symmetrically to get 1000 bp width</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  summits <span class="ot">&lt;-</span> </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    rtracklayer<span class="sc">::</span><span class="fu">import</span>(<span class="fu">paste0</span>(<span class="st">"/home/philipp/AML_Final_Project/data/chip-nexus/"</span>, tf, <span class="st">"/idr-optimal-set.summit.bed"</span>), </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>                           <span class="at">format=</span><span class="st">"bed"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    GenomicRanges<span class="sc">::</span><span class="fu">resize</span>(<span class="at">width=</span>peak_width, <span class="at">fix=</span><span class="st">"center"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>    plyranges<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">TF =</span> tf) <span class="sc">%&gt;%</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    plyranges<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">set =</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">as.character</span>(seqnames) <span class="sc">%in%</span> chrom_list<span class="sc">$</span>tune <span class="sc">~</span> <span class="st">"tune"</span>,</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">as.character</span>(seqnames) <span class="sc">%in%</span> chrom_list<span class="sc">$</span>test <span class="sc">~</span> <span class="st">"test"</span>,</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">as.character</span>(seqnames) <span class="sc">%in%</span> chrom_list<span class="sc">$</span>train <span class="sc">~</span> <span class="st">"train"</span>))</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>  GenomicRanges<span class="sc">::</span><span class="fu">elementMetadata</span>(summits) <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>    GenomicRanges<span class="sc">::</span><span class="fu">elementMetadata</span>(summits),</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>    GenomicRanges<span class="sc">::</span><span class="fu">elementMetadata</span>(narrowPeaks)</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>  summits</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(<span class="at">what=</span>c, <span class="at">args=</span>.)</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a><span class="co"># save peak info as tsv</span></span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a><span class="fu">write_delim</span>(<span class="fu">data.frame</span>(peak_infos) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Region =</span> <span class="fu">paste0</span>(seqnames, <span class="st">":"</span>, start, <span class="st">"-"</span>, end)), </span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>            <span class="at">file=</span><span class="fu">paste0</span>(output_dir, <span class="st">"region_info.tsv"</span>), <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>peak_infos <span class="sc">%&gt;%</span> head</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 6 ranges and 8 metadata columns:
      seqnames              ranges strand |          TF         set        name
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt;
  [1]     chrX 143482559-143483558      * |        Sox2       train        &lt;NA&gt;
  [2]     chrY     4149724-4150723      * |        Sox2       train        &lt;NA&gt;
  [3]     chr9     3001634-3002633      * |        Sox2        test        &lt;NA&gt;
  [4]     chr3 122145078-122146077      * |        Sox2        tune        &lt;NA&gt;
  [5]     chr5   28209565-28210564      * |        Sox2       train        &lt;NA&gt;
  [6]     chr8   86393561-86394560      * |        Sox2        test        &lt;NA&gt;
          score signalValue    pValue    qValue      peak
      &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;
  [1]      1000    65.04316  37093.23  37086.70       145
  [2]      1000    59.28276   2920.20   2914.44       142
  [3]      1000     7.05411   2326.30   2320.59       128
  [4]      1000    25.36805   1518.00   1512.39       407
  [5]      1000    29.15046   1502.75   1497.15       292
  [6]      1000    27.50292   1455.06   1449.46       307
  -------
  seqinfo: 21 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<section id="distribution-of-the-macs2-scores" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="distribution-of-the-macs2-scores"><span class="header-section-number">5.1</span> Distribution of the MACS2 Scores</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> peak_infos <span class="sc">%&gt;%</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fu">log2</span>(qValue), <span class="at">color=</span>TF), <span class="at">alpha=</span><span class="fl">0.4</span>, <span class="at">fill=</span><span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Log2 qValue"</span>, <span class="at">y=</span><span class="st">"Density"</span>, <span class="at">color=</span><span class="st">"TF"</span>) <span class="sc">+</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span>colors) <span class="sc">+</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">"/home/philipp/BPNet/out/figures/macs2_scores_per_tf.pdf"</span>, </span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">plot =</span> p, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="384"></p>
</div>
</div>
</section>
<section id="number-of-peaks-per-tf" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="number-of-peaks-per-tf"><span class="header-section-number">5.2</span> Number of Peaks per TF</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> peak_infos <span class="sc">%&gt;%</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">y=</span>TF, <span class="at">fill=</span>TF), <span class="at">alpha=</span><span class="fl">0.4</span>, <span class="at">width=</span><span class="fl">0.4</span>, <span class="at">color=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Number of Peaks"</span>, <span class="at">y=</span><span class="st">"TF"</span>) <span class="sc">+</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values=</span>colors)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">"/home/philipp/BPNet/out/figures/n_peaks_per_tf.pdf"</span>, </span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">plot =</span> p, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>)</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="384"></p>
</div>
</div>
</section>
</section>
<section id="translate-peak-sequences-to-one-hot-encoding" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Translate Peak Sequences to One-Hot Encoding</h1>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-12_a383127c0566fd48ae5e5c1fe527552f">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>seq_names <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">imap</span>(chrom_list, <span class="cf">function</span>(set_chroms, set_name) {</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  peak_info_subset <span class="ot">&lt;-</span> peak_infos[<span class="fu">seqnames</span>(peak_infos) <span class="sc">%in%</span> set_chroms]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  rnames <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">as.character</span>(peak_info_subset<span class="sc">@</span>seqnames), <span class="st">":"</span>,</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>         peak_info_subset<span class="sc">@</span>ranges<span class="sc">@</span>start, <span class="st">"-"</span>,</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>         peak_info_subset<span class="sc">@</span>ranges<span class="sc">@</span>start <span class="sc">+</span> peak_info_subset<span class="sc">@</span>ranges<span class="sc">@</span>width <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write_lines</span>(rnames, <span class="at">file =</span> <span class="fu">paste0</span>(output_dir, set_name, <span class="st">"_seq_names.txt"</span>))</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  rnames</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-13_78dc5cdce89b099417057b0b08157c09">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(seq_names)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ tune : chr [1:29277] "chr3:122145078-122146077" "chr3:108433158-108434157" "chr2:52071742-52072741" "chr4:141565124-141566123" ...
 $ test : chr [1:27727] "chr9:3001634-3002633" "chr8:86393561-86394560" "chr9:102205227-102206226" "chr1:134534776-134535775" ...
 $ train: chr [1:93904] "chrX:143482559-143483558" "chrY:4149724-4150723" "chr5:28209565-28210564" "chr12:74803793-74804792" ...</code></pre>
</div>
</div>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-14_04cf158726ee2b361c5d72fd1b3108f5">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>one_hot_seqs <span class="ot">&lt;-</span> furrr<span class="sc">::</span><span class="fu">future_imap</span>(chrom_list, <span class="cf">function</span>(set_chroms, set_name) {</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  peak_info_subset <span class="ot">&lt;-</span> peak_infos[<span class="fu">seqnames</span>(peak_infos) <span class="sc">%in%</span> set_chroms]</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  peak_seqs <span class="ot">&lt;-</span> BSgenome<span class="sc">::</span><span class="fu">getSeq</span>(</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    BSgenome.Mmusculus.UCSC.mm10,</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">names=</span><span class="fu">as.character</span>(peak_info_subset<span class="sc">@</span>seqnames),</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">start=</span>(peak_info_subset<span class="sc">@</span>ranges<span class="sc">@</span>start),</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">end=</span>(peak_info_subset<span class="sc">@</span>ranges<span class="sc">@</span>start <span class="sc">+</span> peak_width <span class="sc">-</span> <span class="dv">1</span>),</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">strand=</span><span class="st">"+"</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  peak_seqs<span class="sc">@</span>ranges<span class="sc">@</span>NAMES <span class="ot">&lt;-</span> seq_names[[set_name]]</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># check for correctness here: https://genome.ucsc.edu/cgi-bin/hgTracks?db=mm10</span></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>  Biostrings<span class="sc">::</span><span class="fu">writeXStringSet</span>(<span class="at">x=</span>peak_seqs,</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>                              <span class="at">filepath =</span> <span class="fu">paste0</span>(output_dir, set_name, <span class="st">"_seqs.fa"</span>))</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># one hot encode the sequence</span></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  one_hot_mtx <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="dv">0</span>, <span class="at">nrow=</span><span class="fu">length</span>(peak_seqs)<span class="sc">*</span><span class="dv">4</span>, <span class="at">ncol=</span>peak_width)</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (peak_index <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(peak_seqs)) {</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>    one_hot_mtx[(((peak_index<span class="dv">-1</span>)<span class="sc">*</span><span class="dv">4</span>)<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span>((peak_index)<span class="sc">*</span><span class="dv">4</span>), ] <span class="ot">&lt;-</span></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>      <span class="fu">one_hot_c</span>(<span class="fu">as.character</span>(peak_seqs[[peak_index]]))</span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>  one_hot_mtx</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the output:</p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-15_c8065d0496717b779a560ce4716e6f34">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(one_hot_seqs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ tune : num [1:117108, 1:1000] 1 0 0 0 0 1 0 0 0 0 ...
 $ test : num [1:110908, 1:1000] 0 0 0 1 0 1 0 0 0 0 ...
 $ train: num [1:375616, 1:1000] 0 0 0 1 1 0 0 0 0 1 ...</code></pre>
</div>
</div>
<p>Write the matrices in binary format to disk using <code>reticulate</code>.</p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-16_77922b072d888642cbb03471984a2fb0">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> key, val <span class="kw">in</span> r.one_hot_seqs.items():</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  old_shape <span class="op">=</span> val.shape</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  val <span class="op">=</span> val.reshape(<span class="bu">int</span>(val.shape[<span class="dv">0</span>]<span class="op">/</span><span class="dv">4</span>), <span class="dv">4</span>, <span class="dv">1000</span>).astype(np.float32)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>key<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span>old_shape<span class="sc">}</span><span class="ss"> -&gt; </span><span class="sc">{</span>val<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  save_path <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>r<span class="sc">.</span>output_dir<span class="sc">}{</span>key<span class="sc">}</span><span class="ss">_one_hot_seqs.npy"</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"-&gt; </span><span class="sc">{</span>save_path<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  np.save(<span class="bu">file</span><span class="op">=</span>save_path, arr<span class="op">=</span>val)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tune: (117108, 1000) -&gt; (29277, 4, 1000)
-&gt; /home/philipp/BPNet/input/tune_one_hot_seqs.npy
test: (110908, 1000) -&gt; (27727, 4, 1000)
-&gt; /home/philipp/BPNet/input/test_one_hot_seqs.npy
train: (375616, 1000) -&gt; (93904, 4, 1000)
-&gt; /home/philipp/BPNet/input/train_one_hot_seqs.npy</code></pre>
</div>
</div>
</section>
<section id="extract-the-tf-counts" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Extract the TF Counts</h1>
<p>First we have to merge overlapping peaks from the peak info file. Otherwise we read reads aligning to these overlapping peaks several times.</p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-17_f97b169440df74dfc1fc6b3ffd5596ac">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>peak_infos_reduced <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">reduce</span>(<span class="fu">resize</span>(peak_infos, GenomicRanges<span class="sc">::</span><span class="fu">width</span>(peak_infos <span class="sc">+</span> <span class="dv">1</span>), <span class="st">"start"</span>))</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>peak_infos_reduced</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 85250 ranges and 0 metadata columns:
          seqnames              ranges strand
             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
      [1]     chrX     5704156-5705557      *
      [2]     chrX     6278899-6280002      *
      [3]     chrX     6411759-6412760      *
      [4]     chrX     6415698-6416699      *
      [5]     chrX     6445022-6446023      *
      ...      ...                 ...    ...
  [85246]    chr10 130523572-130524574      *
  [85247]    chr10 130527565-130528566      *
  [85248]    chr10 130535527-130536528      *
  [85249]    chr10 130541495-130543475      *
  [85250]    chr10 130594405-130595406      *
  -------
  seqinfo: 21 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-18_61cd3857fc107251e502e4442be2716a">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Parallel loop over all tfs</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>tf_counts <span class="ot">&lt;-</span> TFs <span class="sc">%&gt;%</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">set_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  furrr<span class="sc">::</span><span class="fu">future_map</span>(<span class="cf">function</span>(tf) {</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># read only from the alignment file in the given peak regions</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># note: make sure that the BAM file is sorted (check for presence of ".bam.bai")</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  alignments <span class="ot">&lt;-</span> <span class="fu">readGAlignments</span>(<span class="fu">paste0</span>(<span class="st">"../data/chip-nexus/"</span>, tf, <span class="st">"/pool_filt.bam"</span>),</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>                                <span class="at">param =</span> <span class="fu">ScanBamParam</span>(<span class="at">which=</span>peak_infos_reduced))</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># split the alignment into pos and neg strand</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>  align_pos <span class="ot">&lt;-</span> alignments[GenomicRanges<span class="sc">::</span><span class="fu">strand</span>(alignments)<span class="sc">==</span><span class="st">"+"</span>]</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>  align_neg <span class="ot">&lt;-</span> alignments[GenomicRanges<span class="sc">::</span><span class="fu">strand</span>(alignments)<span class="sc">==</span><span class="st">"-"</span>]</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># only retain first base pair of each read</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>  align_pos<span class="sc">@</span>cigar <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"1M"</span>, <span class="fu">length</span>(align_pos))</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>  align_neg<span class="sc">@</span>start <span class="ot">&lt;-</span> GenomicAlignments<span class="sc">::</span><span class="fu">end</span>(align_neg)</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>  align_neg<span class="sc">@</span>cigar <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"1M"</span>, <span class="fu">length</span>(align_neg))</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># compute the coverage per base pair</span></span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>  cov_list <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"pos"</span> <span class="ot">=</span> GenomicAlignments<span class="sc">::</span><span class="fu">coverage</span>(align_pos, <span class="at">weight =</span> 1L),</span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"neg"</span> <span class="ot">=</span> GenomicAlignments<span class="sc">::</span><span class="fu">coverage</span>(align_neg, <span class="at">weight =</span> 1L))</span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>  chrom_list <span class="sc">%&gt;%</span> </span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">imap</span>(<span class="cf">function</span>(set_chroms, set_name) {</span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a>      peak_info_subset <span class="ot">&lt;-</span> peak_infos[<span class="fu">seqnames</span>(peak_infos) <span class="sc">%in%</span> set_chroms]</span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a>      <span class="fu">c</span>(<span class="st">"pos"</span>, <span class="st">"neg"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a>        purrr<span class="sc">::</span><span class="fu">set_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a>        purrr<span class="sc">::</span><span class="fu">map</span>(<span class="cf">function</span>(strand) {</span>
<span id="cb27-30"><a href="#cb27-30" aria-hidden="true" tabindex="-1"></a>          mtx <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="dv">0</span>, <span class="at">ncol=</span>peak_width, <span class="at">nrow=</span><span class="fu">length</span>(peak_info_subset))</span>
<span id="cb27-31"><a href="#cb27-31" aria-hidden="true" tabindex="-1"></a>          <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(peak_info_subset)) {</span>
<span id="cb27-32"><a href="#cb27-32" aria-hidden="true" tabindex="-1"></a>            chr <span class="ot">&lt;-</span> <span class="fu">as.character</span>(peak_info_subset[i]<span class="sc">@</span>seqnames)</span>
<span id="cb27-33"><a href="#cb27-33" aria-hidden="true" tabindex="-1"></a>            position_index <span class="ot">&lt;-</span> peak_info_subset[i]<span class="sc">@</span>ranges<span class="sc">@</span>start<span class="sc">:</span>(peak_info_subset[i]<span class="sc">@</span>ranges<span class="sc">@</span>start <span class="sc">+</span> peak_width <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb27-34"><a href="#cb27-34" aria-hidden="true" tabindex="-1"></a>            mtx[i, ] <span class="ot">=</span> <span class="fu">as.numeric</span>(cov_list[[strand]][[chr]][position_index])</span>
<span id="cb27-35"><a href="#cb27-35" aria-hidden="true" tabindex="-1"></a>          }</span>
<span id="cb27-36"><a href="#cb27-36" aria-hidden="true" tabindex="-1"></a>          mtx</span>
<span id="cb27-37"><a href="#cb27-37" aria-hidden="true" tabindex="-1"></a>      })</span>
<span id="cb27-38"><a href="#cb27-38" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb27-39"><a href="#cb27-39" aria-hidden="true" tabindex="-1"></a>  })</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-19_6de4e0a0b245511b9494e396f15170d9">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(tf_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 4
 $ Sox2 :List of 3
  ..$ tune :List of 2
  .. ..$ pos: num [1:29277, 1:1000] 0 0 0 0 0 0 0 0 1 0 ...
  .. ..$ neg: num [1:29277, 1:1000] 0 0 1 0 0 0 0 0 1 0 ...
  ..$ test :List of 2
  .. ..$ pos: num [1:27727, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
  .. ..$ neg: num [1:27727, 1:1000] 0 0 0 0 0 0 0 1 0 0 ...
  ..$ train:List of 2
  .. ..$ pos: num [1:93904, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
  .. ..$ neg: num [1:93904, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
 $ Oct4 :List of 3
  ..$ tune :List of 2
  .. ..$ pos: num [1:29277, 1:1000] 1 1 1 0 0 0 0 0 0 0 ...
  .. ..$ neg: num [1:29277, 1:1000] 1 0 1 0 0 0 0 0 0 0 ...
  ..$ test :List of 2
  .. ..$ pos: num [1:27727, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
  .. ..$ neg: num [1:27727, 1:1000] 0 0 0 0 0 2 0 0 0 0 ...
  ..$ train:List of 2
  .. ..$ pos: num [1:93904, 1:1000] 0 0 0 0 0 0 0 0 2 0 ...
  .. ..$ neg: num [1:93904, 1:1000] 0 0 1 0 0 0 0 0 0 0 ...
 $ Klf4 :List of 3
  ..$ tune :List of 2
  .. ..$ pos: num [1:29277, 1:1000] 0 0 1 0 0 0 0 0 0 0 ...
  .. ..$ neg: num [1:29277, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
  ..$ test :List of 2
  .. ..$ pos: num [1:27727, 1:1000] 0 1 0 0 0 0 0 0 0 0 ...
  .. ..$ neg: num [1:27727, 1:1000] 0 0 0 0 0 3 0 0 0 0 ...
  ..$ train:List of 2
  .. ..$ pos: num [1:93904, 1:1000] 0 0 0 0 0 1 0 0 2 0 ...
  .. ..$ neg: num [1:93904, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
 $ Nanog:List of 3
  ..$ tune :List of 2
  .. ..$ pos: num [1:29277, 1:1000] 1 0 1 0 0 0 0 0 0 0 ...
  .. ..$ neg: num [1:29277, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
  ..$ test :List of 2
  .. ..$ pos: num [1:27727, 1:1000] 0 1 0 1 0 0 0 0 0 0 ...
  .. ..$ neg: num [1:27727, 1:1000] 0 1 0 0 0 1 0 1 0 0 ...
  ..$ train:List of 2
  .. ..$ pos: num [1:93904, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
  .. ..$ neg: num [1:93904, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...</code></pre>
</div>
</div>
<p>Write the matrices in binary format to disk using <code>reticulate</code>.</p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-20_2543552a7690d75f134dbad55be7bdfa">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> tf_name, tf_entry <span class="kw">in</span> r.tf_counts.items():</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(tf_name)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> set_name, set_entry <span class="kw">in</span> tf_entry.items():</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    mtx <span class="op">=</span> np.stack([set_entry[<span class="st">"pos"</span>], set_entry[<span class="st">"neg"</span>]], axis<span class="op">=</span><span class="dv">1</span>).astype(np.float32)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="sc">{</span>set_name<span class="sc">}</span><span class="ss"> shape: </span><span class="sc">{</span>mtx<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    save_path <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>r<span class="sc">.</span>output_dir<span class="sc">}{</span>tf_name<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>set_name<span class="sc">}</span><span class="ss">_counts.npy"</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">-&gt; </span><span class="sc">{</span>save_path<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>    np.save(<span class="bu">file</span><span class="op">=</span>save_path, arr<span class="op">=</span>mtx)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sox2
    tune shape: (29277, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Sox2/tune_counts.npy
    test shape: (27727, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Sox2/test_counts.npy
    train shape: (93904, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Sox2/train_counts.npy
Oct4
    tune shape: (29277, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Oct4/tune_counts.npy
    test shape: (27727, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Oct4/test_counts.npy
    train shape: (93904, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Oct4/train_counts.npy
Klf4
    tune shape: (29277, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Klf4/tune_counts.npy
    test shape: (27727, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Klf4/test_counts.npy
    train shape: (93904, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Klf4/train_counts.npy
Nanog
    tune shape: (29277, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Nanog/tune_counts.npy
    test shape: (27727, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Nanog/test_counts.npy
    train shape: (93904, 2, 1000)
    -&gt; /home/philipp/BPNet/input/Nanog/train_counts.npy</code></pre>
</div>
</div>
<section id="examples-highest-scoring-peaks" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="examples-highest-scoring-peaks"><span class="header-section-number">7.1</span> Examples: Highest Scoring Peaks</h2>
<details>
<summary>
Click to expand
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>test_df <span class="ot">&lt;-</span> peak_infos <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(set<span class="sc">==</span><span class="st">"train"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Region =</span> <span class="fu">paste0</span>(seqnames, <span class="st">":"</span>, start, <span class="st">"-"</span>, end)) <span class="sc">%&gt;%</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(TF) <span class="sc">%&gt;%</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(<span class="at">order_by=</span>qValue, <span class="at">n=</span><span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Region, TF, qValue)</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>purrr<span class="sc">::</span><span class="fu">walk</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(test_df), <span class="cf">function</span>(i) {</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>  test_instance <span class="ot">&lt;-</span> <span class="fu">unlist</span>(test_df[i, ])</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map_dfr</span>(TFs, <span class="cf">function</span>(tf){</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">plus =</span> tf_counts[[tf]]<span class="sc">$</span>train<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ],</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">minus =</span> <span class="sc">-</span>tf_counts[[tf]]<span class="sc">$</span>train<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ]) <span class="sc">%&gt;%</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">TF =</span> tf, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span> </span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="fu">c</span>(minus, plus)) <span class="sc">%&gt;%</span></span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>value, <span class="at">color=</span>name)) <span class="sc">+</span></span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>TF, <span class="at">scales=</span><span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>() <span class="sc">+</span></span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Relative Position [bp]"</span>, <span class="at">y=</span><span class="st">"Counts"</span>, <span class="at">color=</span><span class="st">"Strand"</span>, </span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"Peak "</span>, test_instance[<span class="st">"Region"</span>], <span class="st">" | "</span>, test_instance[<span class="st">"TF"</span>], </span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" | "</span>, test_instance[<span class="st">"qValue"</span>])) <span class="ot">-&gt;</span> p</span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(p)</span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-9.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-10.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-11.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-12.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-13.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-14.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-15.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-16.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-17.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-18.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-19.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-21-20.png" class="img-fluid" width="672"></p>
</div>
</div>
</details>
</section>
<section id="examples-random-peaks" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="examples-random-peaks"><span class="header-section-number">7.2</span> Examples: Random Peaks</h2>
<details>
<summary>
Click to expand
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>test_df <span class="ot">&lt;-</span> peak_infos <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(set<span class="sc">==</span><span class="st">"train"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Region =</span> <span class="fu">paste0</span>(seqnames, <span class="st">":"</span>, start, <span class="st">"-"</span>, end)) <span class="sc">%&gt;%</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(TF) <span class="sc">%&gt;%</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_sample</span>(<span class="at">n=</span><span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Region, TF, qValue)</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>purrr<span class="sc">::</span><span class="fu">walk</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(test_df), <span class="cf">function</span>(i) {</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>  test_instance <span class="ot">&lt;-</span> <span class="fu">unlist</span>(test_df[i, ])</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map_dfr</span>(TFs, <span class="cf">function</span>(tf){</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">plus =</span> tf_counts[[tf]]<span class="sc">$</span>train<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ],</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>                 <span class="at">minus =</span> <span class="sc">-</span>tf_counts[[tf]]<span class="sc">$</span>train<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ]) <span class="sc">%&gt;%</span></span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">TF =</span> tf, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span> </span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="fu">c</span>(minus, plus)) <span class="sc">%&gt;%</span></span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>value, <span class="at">color=</span>name)) <span class="sc">+</span></span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>TF, <span class="at">scales=</span><span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>() <span class="sc">+</span></span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Relative Position [bp]"</span>, <span class="at">y=</span><span class="st">"Counts"</span>, <span class="at">color=</span><span class="st">"Strand"</span>, </span>
<span id="cb33-24"><a href="#cb33-24" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"Peak "</span>, test_instance[<span class="st">"Region"</span>], <span class="st">" | "</span>, test_instance[<span class="st">"TF"</span>], </span>
<span id="cb33-25"><a href="#cb33-25" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" | "</span>, test_instance[<span class="st">"qValue"</span>])) <span class="ot">-&gt;</span> p</span>
<span id="cb33-26"><a href="#cb33-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(p)</span>
<span id="cb33-27"><a href="#cb33-27" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-9.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-10.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-11.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-12.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-13.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-14.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-15.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-16.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-17.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-18.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-19.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-22-20.png" class="img-fluid" width="672"></p>
</div>
</div>
</details>
</section>
</section>
<section id="extract-the-control-patchcap-counts" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Extract the Control (Patchcap) Counts</h1>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-23_6df2e88cbd817e252c08edc8e0bf387d">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>patchcap_input <span class="ot">&lt;-</span> </span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"pos"</span>, <span class="st">"neg"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">set_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map</span>(<span class="cf">function</span>(strand){</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  rtracklayer<span class="sc">::</span><span class="fu">import.bw</span>(<span class="fu">paste0</span>(<span class="st">"../data/chip-nexus/patchcap/counts."</span>, strand, <span class="st">".bw"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    GenomicRanges<span class="sc">::</span><span class="fu">coverage</span>(., <span class="at">weight =</span> <span class="st">"score"</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>ctrl_counts <span class="ot">&lt;-</span> furrr<span class="sc">::</span><span class="fu">future_imap</span>(chrom_list, <span class="cf">function</span>(set_chroms, set_name) {</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>  peak_info_subset <span class="ot">&lt;-</span> peak_infos[<span class="fu">seqnames</span>(peak_infos) <span class="sc">%in%</span> set_chroms]</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"pos"</span>, <span class="st">"neg"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">set_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">map</span>(<span class="cf">function</span>(strand) {</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>      mtx <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="dv">0</span>, <span class="at">ncol=</span>peak_width, <span class="at">nrow=</span><span class="fu">length</span>(peak_info_subset))</span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> (seq_index <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(mtx)) {</span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>        chrom_index <span class="ot">&lt;-</span> <span class="fu">as.character</span>(peak_info_subset[seq_index]<span class="sc">@</span>seqnames)</span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>        position_index <span class="ot">&lt;-</span> peak_info_subset[seq_index]<span class="sc">@</span>ranges<span class="sc">@</span>start<span class="sc">:</span>(peak_info_subset[seq_index]<span class="sc">@</span>ranges<span class="sc">@</span>start<span class="sc">+</span>peak_width<span class="dv">-1</span>)</span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>        mtx[seq_index, ] <span class="ot">=</span> <span class="fu">as.numeric</span>(patchcap_input[[strand]][[chrom_index]][position_index])</span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>      mtx</span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-24_cec5fa89f332c9a9fc70ada09e333183">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(ctrl_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ tune :List of 2
  ..$ pos: num [1:29277, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
  ..$ neg: num [1:29277, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
 $ test :List of 2
  ..$ pos: num [1:27727, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
  ..$ neg: num [1:27727, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...
 $ train:List of 2
  ..$ pos: num [1:93904, 1:1000] 0 0 0 0 0 0 0 0 1 0 ...
  ..$ neg: num [1:93904, 1:1000] 0 0 0 0 0 0 0 0 0 0 ...</code></pre>
</div>
</div>
<p>Write the matrices in binary format to disk using <code>reticulate</code></p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-25_9b31255cc28cd6b8e3adf49a822f2006">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> set_name, set_entry <span class="kw">in</span> r.ctrl_counts.items():</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  mtx <span class="op">=</span> np.stack([set_entry[<span class="st">"pos"</span>], set_entry[<span class="st">"neg"</span>]], axis<span class="op">=</span><span class="dv">1</span>).astype(np.float32)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="sc">{</span>set_name<span class="sc">}</span><span class="ss"> shape: </span><span class="sc">{</span>mtx<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  save_path <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>r<span class="sc">.</span>output_dir<span class="sc">}</span><span class="ss">patchcap/</span><span class="sc">{</span>set_name<span class="sc">}</span><span class="ss">_counts.npy"</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss">-&gt; </span><span class="sc">{</span>save_path<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>  np.save(<span class="bu">file</span><span class="op">=</span>save_path, arr<span class="op">=</span>mtx)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    tune shape: (29277, 2, 1000)
    -&gt; /home/philipp/BPNet/input/patchcap/tune_counts.npy
    test shape: (27727, 2, 1000)
    -&gt; /home/philipp/BPNet/input/patchcap/test_counts.npy
    train shape: (93904, 2, 1000)
    -&gt; /home/philipp/BPNet/input/patchcap/train_counts.npy</code></pre>
</div>
</div>
<section id="examples-highest-scoring-peaks-with-bias" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="examples-highest-scoring-peaks-with-bias"><span class="header-section-number">8.1</span> Examples: Highest Scoring Peaks with Bias</h2>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-26_e08f1296fd6968b2ddfbe3f2395f93a8">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>test_instance <span class="ot">&lt;-</span> peak_infos <span class="sc">%&gt;%</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(set<span class="sc">==</span><span class="st">"test"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(seqnames<span class="sc">==</span><span class="st">"chr1"</span>, start <span class="sc">&gt;=</span> <span class="dv">180924752-1000</span>, end <span class="sc">&lt;=</span> <span class="dv">180925152</span><span class="sc">+</span><span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Region =</span> <span class="fu">paste0</span>(seqnames, <span class="st">":"</span>, start, <span class="st">"-"</span>, end)) <span class="sc">%&gt;%</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(Region, TF, qValue) <span class="sc">%&gt;%</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>  .[<span class="dv">1</span>, ]</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(test_instance)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                    Region   TF   qValue
1 chr1:180924435-180925434 Sox2 436.0653</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map_dfr</span>(TFs, <span class="cf">function</span>(tf){</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">plus =</span> tf_counts[[tf]]<span class="sc">$</span>test<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>test), ],</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">minus =</span> <span class="sc">-</span>tf_counts[[tf]]<span class="sc">$</span>test<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>test), ]) <span class="sc">%&gt;%</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">TF =</span> tf, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span> </span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbind</span>(</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">plus =</span> ctrl_counts<span class="sc">$</span>test<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>test), ],</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">minus =</span> <span class="sc">-</span>ctrl_counts<span class="sc">$</span>test<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>test), ],</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">TF =</span> <span class="st">"patchcap"</span>, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="fu">c</span>(minus, plus)) <span class="sc">%&gt;%</span></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">TF =</span> <span class="fu">factor</span>(TF, <span class="at">levels=</span><span class="fu">names</span>(colors))) <span class="sc">%&gt;%</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>value, <span class="at">color=</span>TF, <span class="at">alpha=</span>name), <span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>TF, <span class="at">ncol=</span><span class="dv">1</span>, <span class="at">scales=</span><span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Relative Position [bp]"</span>, <span class="at">y=</span><span class="st">"Counts"</span>, <span class="at">color=</span><span class="st">"Strand"</span>, </span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span>test_instance[<span class="st">"Region"</span>]) <span class="sc">+</span></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>  <span class="co">#scale_color_manual(values=c("minus"="darkred", "plus"="forestgreen")) +</span></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span>colors) <span class="sc">+</span></span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_alpha_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"plus"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"minus"</span> <span class="ot">=</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">10</span>, <span class="at">hjust=</span><span class="fl">0.5</span>), <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="cn">NA</span>))</span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">"/home/philipp/BPNet/out/figures/example_high_q.pdf"</span>, </span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a>       <span class="at">plot =</span> p, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">8</span>)</span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> </span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="384"></p>
</div>
</div>
<details>
<summary>
Click to view more plots
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>test_df <span class="ot">&lt;-</span> peak_infos <span class="sc">%&gt;%</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(set<span class="sc">==</span><span class="st">"train"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Region =</span> <span class="fu">paste0</span>(seqnames, <span class="st">":"</span>, start, <span class="st">"-"</span>, end)) <span class="sc">%&gt;%</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(TF) <span class="sc">%&gt;%</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(<span class="at">order_by=</span>qValue, <span class="at">n=</span><span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Region, TF, qValue)</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>purrr<span class="sc">::</span><span class="fu">walk</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(test_df), <span class="cf">function</span>(i) {</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>  test_instance <span class="ot">&lt;-</span> <span class="fu">unlist</span>(test_df[i, ])</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map_dfr</span>(TFs, <span class="cf">function</span>(tf){</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>                 <span class="at">plus =</span> tf_counts[[tf]]<span class="sc">$</span>train<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ],</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">minus =</span> <span class="sc">-</span>tf_counts[[tf]]<span class="sc">$</span>train<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ]) <span class="sc">%&gt;%</span></span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">TF =</span> tf, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span> </span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbind</span>(</span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>                   <span class="at">plus =</span> ctrl_counts<span class="sc">$</span>train<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ],</span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a>                   <span class="at">minus =</span> <span class="sc">-</span>ctrl_counts<span class="sc">$</span>train<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ],</span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>                   <span class="at">TF =</span> <span class="st">"Bias"</span>, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="fu">c</span>(minus, plus)) <span class="sc">%&gt;%</span></span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>value, <span class="at">color=</span>name)) <span class="sc">+</span></span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>TF, <span class="at">scales=</span><span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>() <span class="sc">+</span></span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Relative Position [bp]"</span>, <span class="at">y=</span><span class="st">"Counts"</span>, <span class="at">color=</span><span class="st">"Strand"</span>, </span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"Peak "</span>, test_instance[<span class="st">"Region"</span>], <span class="st">" | "</span>, test_instance[<span class="st">"TF"</span>], </span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" | "</span>, test_instance[<span class="st">"qValue"</span>])) <span class="ot">-&gt;</span> p</span>
<span id="cb42-30"><a href="#cb42-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(p)</span>
<span id="cb42-31"><a href="#cb42-31" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-9.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-10.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-11.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-12.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-13.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-14.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-15.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-16.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-17.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-18.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-19.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-27-20.png" class="img-fluid" width="672"></p>
</div>
</div>
</details></section>
<section id="examples-random-peaks-with-bias" class="level2" data-number="8.2">
<h2 data-number="8.2" class="anchored" data-anchor-id="examples-random-peaks-with-bias"><span class="header-section-number">8.2</span> Examples: Random Peaks with Bias</h2>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-28_ef3d11144ea6f1892da3c9320054a175">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>test_instance <span class="ot">&lt;-</span> peak_infos <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(set<span class="sc">==</span><span class="st">"test"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(seqnames<span class="sc">==</span><span class="st">"chr1"</span>, start <span class="sc">&gt;=</span> <span class="dv">4000</span>, end <span class="sc">&lt;=</span> <span class="dv">100000000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Region =</span> <span class="fu">paste0</span>(seqnames, <span class="st">":"</span>, start, <span class="st">"-"</span>, end)) <span class="sc">%&gt;%</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(Region, TF, qValue) <span class="sc">%&gt;%</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  .[<span class="dv">1000</span>, ]</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(test_instance)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     Region   TF   qValue
1000 chr1:58392603-58393602 Oct4 21.10429</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map_dfr</span>(TFs, <span class="cf">function</span>(tf){</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">plus =</span> tf_counts[[tf]]<span class="sc">$</span>test<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>test), ],</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">minus =</span> <span class="sc">-</span>tf_counts[[tf]]<span class="sc">$</span>test<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>test), ]) <span class="sc">%&gt;%</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">TF =</span> tf, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span> </span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbind</span>(</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">plus =</span> ctrl_counts<span class="sc">$</span>test<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>test), ],</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">minus =</span> <span class="sc">-</span>ctrl_counts<span class="sc">$</span>test<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>test), ],</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">TF =</span> <span class="st">"patchcap"</span>, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="fu">c</span>(minus, plus)) <span class="sc">%&gt;%</span></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">TF =</span> <span class="fu">factor</span>(TF, <span class="at">levels=</span><span class="fu">names</span>(colors))) <span class="sc">%&gt;%</span></span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>value, <span class="at">color=</span>TF, <span class="at">alpha=</span>name), <span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>TF, <span class="at">ncol=</span><span class="dv">1</span>, <span class="at">scales=</span><span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Relative Position [bp]"</span>, <span class="at">y=</span><span class="st">"Counts"</span>, <span class="at">color=</span><span class="st">"Strand"</span>, </span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span>test_instance[<span class="st">"Region"</span>]) <span class="sc">+</span></span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>  <span class="co">#scale_color_manual(values=c("minus"="darkred", "plus"="forestgreen")) +</span></span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span>colors) <span class="sc">+</span></span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_alpha_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"plus"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"minus"</span> <span class="ot">=</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">10</span>, <span class="at">hjust=</span><span class="fl">0.5</span>), <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="cn">NA</span>))</span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">"/home/philipp/BPNet/out/figures/example_low_q.pdf"</span>, </span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a>       <span class="at">plot =</span> p, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">8</span>)</span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> </span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="384"></p>
</div>
</div>
<details>
<summary>
Click to view more plots
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>test_df <span class="ot">&lt;-</span> peak_infos <span class="sc">%&gt;%</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(set<span class="sc">==</span><span class="st">"train"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Region =</span> <span class="fu">paste0</span>(seqnames, <span class="st">":"</span>, start, <span class="st">"-"</span>, end)) <span class="sc">%&gt;%</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(TF) <span class="sc">%&gt;%</span></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_sample</span>(<span class="at">n=</span><span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Region, TF, qValue)</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>purrr<span class="sc">::</span><span class="fu">walk</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(test_df), <span class="cf">function</span>(i) {</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>  test_instance <span class="ot">&lt;-</span> <span class="fu">unlist</span>(test_df[i, ])</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map_dfr</span>(TFs, <span class="cf">function</span>(tf){</span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">plus =</span> tf_counts[[tf]]<span class="sc">$</span>train<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ],</span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a>                 <span class="at">minus =</span> <span class="sc">-</span>tf_counts[[tf]]<span class="sc">$</span>train<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ]) <span class="sc">%&gt;%</span></span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">TF =</span> tf, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb46-17"><a href="#cb46-17" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span> </span>
<span id="cb46-18"><a href="#cb46-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbind</span>(</span>
<span id="cb46-19"><a href="#cb46-19" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">position=</span><span class="sc">-</span><span class="dv">499</span><span class="sc">:</span><span class="dv">500</span>, </span>
<span id="cb46-20"><a href="#cb46-20" aria-hidden="true" tabindex="-1"></a>                   <span class="at">plus =</span> ctrl_counts<span class="sc">$</span>train<span class="sc">$</span>pos[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ],</span>
<span id="cb46-21"><a href="#cb46-21" aria-hidden="true" tabindex="-1"></a>                   <span class="at">minus =</span> <span class="sc">-</span>ctrl_counts<span class="sc">$</span>train<span class="sc">$</span>neg[<span class="fu">match</span>(test_instance[<span class="st">"Region"</span>], seq_names<span class="sc">$</span>train), ],</span>
<span id="cb46-22"><a href="#cb46-22" aria-hidden="true" tabindex="-1"></a>                   <span class="at">TF =</span> <span class="st">"Bias"</span>, <span class="at">p_name =</span> test_instance[<span class="st">"Region"</span>])</span>
<span id="cb46-23"><a href="#cb46-23" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb46-24"><a href="#cb46-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="fu">c</span>(minus, plus)) <span class="sc">%&gt;%</span></span>
<span id="cb46-25"><a href="#cb46-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb46-26"><a href="#cb46-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>value, <span class="at">color=</span>name)) <span class="sc">+</span></span>
<span id="cb46-27"><a href="#cb46-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>TF, <span class="at">scales=</span><span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb46-28"><a href="#cb46-28" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>() <span class="sc">+</span></span>
<span id="cb46-29"><a href="#cb46-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Relative Position [bp]"</span>, <span class="at">y=</span><span class="st">"Counts"</span>, <span class="at">color=</span><span class="st">"Strand"</span>, </span>
<span id="cb46-30"><a href="#cb46-30" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"Peak "</span>, test_instance[<span class="st">"Region"</span>], <span class="st">" | "</span>, test_instance[<span class="st">"TF"</span>], </span>
<span id="cb46-31"><a href="#cb46-31" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" | "</span>, test_instance[<span class="st">"qValue"</span>])) <span class="ot">-&gt;</span> p</span>
<span id="cb46-32"><a href="#cb46-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(p)</span>
<span id="cb46-33"><a href="#cb46-33" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-9.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-10.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-11.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-12.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-13.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-14.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-15.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-16.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-17.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-18.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-19.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-29-20.png" class="img-fluid" width="672"></p>
</div>
</div>
</details></section>
</section>
<section id="figure-1e" class="level1" data-number="9">
<h1 data-number="9"><span class="header-section-number">9</span> Figure 1e</h1>
<section id="from-raw-data" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="from-raw-data"><span class="header-section-number">9.1</span> From Raw Data</h2>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-30_c032132aaf827fd97f40ee96ad84f197">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>roi <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">"seqname"</span><span class="ot">=</span><span class="st">"chr1"</span>, <span class="st">"start"</span><span class="ot">=</span><span class="dv">180924752</span>, <span class="st">"end"</span><span class="ot">=</span><span class="dv">180925152</span>)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>TFs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Oct4"</span>, <span class="st">"Sox2"</span>, <span class="st">"Nanog"</span>, <span class="st">"Klf4"</span>)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map_dfr</span>(TFs, <span class="cf">function</span>(tf) {</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>    alignments <span class="ot">&lt;-</span> <span class="fu">readGAlignments</span>(<span class="fu">paste0</span>(<span class="st">"../data/chip-nexus/"</span>, tf, <span class="st">"/pool_filt.bam"</span>),</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">param =</span> <span class="fu">ScanBamParam</span>(<span class="at">which=</span><span class="fu">GRanges</span>(<span class="fu">paste0</span>(roi<span class="sc">$</span>seqname, <span class="st">":"</span>, roi<span class="sc">$</span>start, <span class="st">"-"</span>, roi<span class="sc">$</span>end))))</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>    align_pos <span class="ot">&lt;-</span> alignments[<span class="fu">strand</span>(alignments)<span class="sc">==</span><span class="st">"+"</span>]</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>    align_neg <span class="ot">&lt;-</span> alignments[<span class="fu">strand</span>(alignments)<span class="sc">==</span><span class="st">"-"</span>]</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>    align_pos<span class="sc">@</span>cigar <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"1M"</span>, <span class="fu">length</span>(align_pos))</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>    align_neg<span class="sc">@</span>start <span class="ot">&lt;-</span> GenomicAlignments<span class="sc">::</span><span class="fu">end</span>(align_neg)</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>    align_neg<span class="sc">@</span>cigar <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"1M"</span>, <span class="fu">length</span>(align_neg))</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">pos=</span><span class="fu">as.numeric</span>(<span class="fu">coverage</span>(align_pos)<span class="sc">$</span>chr1[roi<span class="sc">$</span>start<span class="sc">:</span>roi<span class="sc">$</span>end]),</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>                   <span class="at">neg=</span><span class="sc">-</span><span class="fu">as.numeric</span>(<span class="fu">coverage</span>(align_neg)<span class="sc">$</span>chr1[roi<span class="sc">$</span>start<span class="sc">:</span>roi<span class="sc">$</span>end]),</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>                   <span class="at">position=</span>roi<span class="sc">$</span>start<span class="sc">:</span>roi<span class="sc">$</span>end,</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>                   <span class="at">TF=</span>tf)</span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="fu">c</span>(<span class="st">"pos"</span>, <span class="st">"neg"</span>))</span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,208 x 4
    position TF    name  value
       &lt;int&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;
 1 180924752 Oct4  pos       0
 2 180924752 Oct4  neg       0
 3 180924753 Oct4  pos       2
 4 180924753 Oct4  neg       0
 5 180924754 Oct4  pos       2
 6 180924754 Oct4  neg       0
 7 180924755 Oct4  pos       0
 8 180924755 Oct4  neg       0
 9 180924756 Oct4  pos       0
10 180924756 Oct4  neg       0
# ... with 3,198 more rows
# i Use `print(n = ...)` to see more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">TF =</span> <span class="fu">factor</span>(TF, <span class="at">levels=</span><span class="fu">c</span>(TFs))) <span class="sc">%&gt;%</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>value, <span class="at">col=</span>name)) <span class="sc">+</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>TF, <span class="at">ncol=</span><span class="dv">1</span>, <span class="at">scales=</span><span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Position"</span>, <span class="at">y=</span><span class="st">"Counts"</span>, <span class="at">col=</span><span class="st">"Strand"</span>) <span class="sc">+</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="from-processed-data" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="from-processed-data"><span class="header-section-number">9.2</span> From Processed Data</h2>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-32_570723a7cd36cae03b0d016ea623fe84">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>roi <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">"seqname"</span><span class="ot">=</span><span class="st">"chr1"</span>, <span class="st">"start"</span><span class="ot">=</span><span class="dv">180924752</span>, <span class="st">"end"</span><span class="ot">=</span><span class="dv">180925152</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>roi_adjusted <span class="ot">&lt;-</span> roi</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>roi_adjusted<span class="sc">$</span>start <span class="ot">=</span> roi_adjusted<span class="sc">$</span>start<span class="dv">-1000</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>roi_adjusted<span class="sc">$</span>end <span class="ot">=</span> roi_adjusted<span class="sc">$</span>end<span class="sc">+</span><span class="dv">1000</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>TFs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Oct4"</span>, <span class="st">"Sox2"</span>, <span class="st">"Nanog"</span>, <span class="st">"Klf4"</span>)</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> </span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map_dfr</span>(TFs, <span class="cf">function</span>(tf) {</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>    cov_list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">"pos"</span> <span class="ot">=</span> tf_counts[[tf]]<span class="sc">$</span>test<span class="sc">$</span>pos,</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"neg"</span> <span class="ot">=</span> tf_counts[[tf]]<span class="sc">$</span>test<span class="sc">$</span>neg)</span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>    rnames <span class="ot">&lt;-</span> seq_names<span class="sc">$</span>test</span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a>    gr_rnames <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(rnames)</span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a>    bool_vec <span class="ot">&lt;-</span> </span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a>      (<span class="fu">as.character</span>(gr_rnames<span class="sc">@</span>seqnames) <span class="sc">==</span> roi_adjusted<span class="sc">$</span>seqname <span class="sc">&amp;</span></span>
<span id="cb50-16"><a href="#cb50-16" aria-hidden="true" tabindex="-1"></a>      <span class="fu">start</span>(gr_rnames<span class="sc">@</span>ranges) <span class="sc">&gt;=</span> roi_adjusted<span class="sc">$</span>start <span class="sc">&amp;</span></span>
<span id="cb50-17"><a href="#cb50-17" aria-hidden="true" tabindex="-1"></a>      <span class="fu">end</span>(gr_rnames<span class="sc">@</span>ranges) <span class="sc">&lt;=</span> roi_adjusted<span class="sc">$</span>end)</span>
<span id="cb50-18"><a href="#cb50-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb50-19"><a href="#cb50-19" aria-hidden="true" tabindex="-1"></a>    peak_index <span class="ot">&lt;-</span> <span class="fu">which</span>(bool_vec)[<span class="dv">1</span>]</span>
<span id="cb50-20"><a href="#cb50-20" aria-hidden="true" tabindex="-1"></a>    peak_info <span class="ot">&lt;-</span> gr_rnames[peak_index]</span>
<span id="cb50-21"><a href="#cb50-21" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb50-22"><a href="#cb50-22" aria-hidden="true" tabindex="-1"></a>    diff <span class="ot">&lt;-</span> <span class="dv">180924752</span> <span class="sc">-</span> <span class="fu">start</span>(peak_info<span class="sc">@</span>ranges) <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb50-23"><a href="#cb50-23" aria-hidden="true" tabindex="-1"></a>    w <span class="ot">&lt;-</span> <span class="dv">400</span></span>
<span id="cb50-24"><a href="#cb50-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb50-25"><a href="#cb50-25" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">pos=</span>cov_list<span class="sc">$</span>pos[peak_index, diff<span class="sc">:</span>(diff<span class="sc">+</span>w)],</span>
<span id="cb50-26"><a href="#cb50-26" aria-hidden="true" tabindex="-1"></a>                   <span class="at">neg=</span><span class="sc">-</span>cov_list<span class="sc">$</span>neg[peak_index, diff<span class="sc">:</span>(diff<span class="sc">+</span>w)],</span>
<span id="cb50-27"><a href="#cb50-27" aria-hidden="true" tabindex="-1"></a>                   <span class="at">position=</span><span class="dv">0</span><span class="sc">:</span><span class="dv">400</span>,</span>
<span id="cb50-28"><a href="#cb50-28" aria-hidden="true" tabindex="-1"></a>                   <span class="at">TF=</span>tf)</span>
<span id="cb50-29"><a href="#cb50-29" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span></span>
<span id="cb50-30"><a href="#cb50-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="fu">c</span>(<span class="st">"pos"</span>, <span class="st">"neg"</span>))</span>
<span id="cb50-31"><a href="#cb50-31" aria-hidden="true" tabindex="-1"></a>df2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,208 x 4
   position TF    name  value
      &lt;int&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;
 1        0 Oct4  pos       0
 2        0 Oct4  neg       0
 3        1 Oct4  pos       2
 4        1 Oct4  neg       0
 5        2 Oct4  pos       2
 6        2 Oct4  neg       0
 7        3 Oct4  pos       0
 8        3 Oct4  neg       0
 9        4 Oct4  pos       0
10        4 Oct4  neg       0
# ... with 3,198 more rows
# i Use `print(n = ...)` to see more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>df2 <span class="sc">%&gt;%</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">TF =</span> <span class="fu">factor</span>(TF, <span class="at">levels=</span><span class="fu">c</span>(TFs))) <span class="sc">%&gt;%</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>value, <span class="at">col=</span>name)) <span class="sc">+</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>TF, <span class="at">ncol=</span><span class="dv">1</span>, <span class="at">scales=</span><span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Position"</span>, <span class="at">y=</span><span class="st">"Counts"</span>, <span class="at">col=</span><span class="st">"Strand"</span>) <span class="sc">+</span></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Compare.</p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-34_e0dca49a242f94122ac21f65c52395ef">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(df2<span class="sc">$</span>value <span class="sc">==</span> df<span class="sc">$</span>value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(df2<span class="sc">$</span>TF <span class="sc">==</span> df<span class="sc">$</span>TF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(df2<span class="sc">$</span>name <span class="sc">==</span> df<span class="sc">$</span>name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
</section>
</section>
<section id="oct4-sox2-motif" class="level1" data-number="10">
<h1 data-number="10"><span class="header-section-number">10</span> Oct4-Sox2 Motif</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>motif <span class="ot">&lt;-</span> TFBSTools<span class="sc">::</span><span class="fu">getMatrixByID</span>(JASPAR2020, <span class="at">ID =</span> <span class="st">"MA0142.1"</span>)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>seqLogo<span class="sc">::</span><span class="fu">seqLogo</span>(motif<span class="sc">@</span>profileMatrix <span class="sc">/</span> <span class="fu">colSums</span>(motif<span class="sc">@</span>profileMatrix))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-36_84133ac069e9adb9b03664ef935ec98a">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>w <span class="ot">=</span> <span class="dv">400</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>motif_matches <span class="ot">&lt;-</span> motifmatchr<span class="sc">::</span><span class="fu">matchMotifs</span>(motif, peak_infos, <span class="at">genome =</span> <span class="st">"mm10"</span>, <span class="at">out=</span><span class="st">"position"</span>)[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>  GenomicRanges<span class="sc">::</span><span class="fu">resize</span>(<span class="at">width=</span>w, <span class="at">fix=</span><span class="st">"center"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>  plyranges<span class="sc">::</span><span class="fu">filter</span>(score <span class="sc">&gt;</span> <span class="fl">20.8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>  plyranges<span class="sc">::</span><span class="fu">arrange</span>(score)</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(motif_matches, <span class="st">"/home/philipp/BPNet/input/oct4_sox2_matches.rds"</span>)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>motif_matches</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 408 ranges and 1 metadata column:
        seqnames            ranges strand |     score
           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
    [1]     chr4 84528359-84528758      + |   20.8166
    [2]     chr4 84528359-84528758      + |   20.8166
    [3]    chr12 86188063-86188462      - |   20.8166
    [4]    chr16 23504381-23504780      + |   20.8166
    [5]     chr9 14619412-14619811      + |   20.8166
    ...      ...               ...    ... .       ...
  [404]    chr15 75270453-75270852      + |   23.4055
  [405]    chr15 80618982-80619381      - |   23.4055
  [406]     chr7 35617437-35617836      - |   23.4055
  [407]     chr8 76863085-76863484      + |   23.4055
  [408]     chr1 75997151-75997550      + |   23.4055
  -------
  seqinfo: 21 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-37_b92fb097c5f7f4309e27a377178b429e">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>motif_matches_reduced <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">reduce</span>(<span class="fu">resize</span>(motif_matches, GenomicRanges<span class="sc">::</span><span class="fu">width</span>(motif_matches <span class="sc">+</span> <span class="dv">1</span>), <span class="st">"start"</span>))</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="do">### Parallel loop over all tfs</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>motif_counts <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Oct4"</span>, <span class="st">"Sox2"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">set_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>  furrr<span class="sc">::</span><span class="fu">future_map</span>(<span class="cf">function</span>(tf) {</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># read only from the alignment file in the given peak regions</span></span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># note: make sure that the BAM file is sorted (check for presence of ".bam.bai")</span></span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>  alignments <span class="ot">&lt;-</span> <span class="fu">readGAlignments</span>(<span class="fu">paste0</span>(<span class="st">"../data/chip-nexus/"</span>, tf, <span class="st">"/pool_filt.bam"</span>),</span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>                                <span class="at">param =</span> <span class="fu">ScanBamParam</span>(<span class="at">which=</span>motif_matches_reduced))</span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># split the alignment into pos and neg strand</span></span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a>  align_pos <span class="ot">&lt;-</span> alignments[GenomicRanges<span class="sc">::</span><span class="fu">strand</span>(alignments)<span class="sc">==</span><span class="st">"+"</span>]</span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a>  align_neg <span class="ot">&lt;-</span> alignments[GenomicRanges<span class="sc">::</span><span class="fu">strand</span>(alignments)<span class="sc">==</span><span class="st">"-"</span>]</span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-17"><a href="#cb62-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># only retain first base pair of each read</span></span>
<span id="cb62-18"><a href="#cb62-18" aria-hidden="true" tabindex="-1"></a>  align_pos<span class="sc">@</span>cigar <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"1M"</span>, <span class="fu">length</span>(align_pos))</span>
<span id="cb62-19"><a href="#cb62-19" aria-hidden="true" tabindex="-1"></a>  align_neg<span class="sc">@</span>start <span class="ot">&lt;-</span> GenomicAlignments<span class="sc">::</span><span class="fu">end</span>(align_neg)</span>
<span id="cb62-20"><a href="#cb62-20" aria-hidden="true" tabindex="-1"></a>  align_neg<span class="sc">@</span>cigar <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"1M"</span>, <span class="fu">length</span>(align_neg))</span>
<span id="cb62-21"><a href="#cb62-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-22"><a href="#cb62-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># compute the coverage per base pair</span></span>
<span id="cb62-23"><a href="#cb62-23" aria-hidden="true" tabindex="-1"></a>  cov_list <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"pos"</span> <span class="ot">=</span> GenomicAlignments<span class="sc">::</span><span class="fu">coverage</span>(align_pos, <span class="at">weight =</span> 1L),</span>
<span id="cb62-24"><a href="#cb62-24" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"neg"</span> <span class="ot">=</span> GenomicAlignments<span class="sc">::</span><span class="fu">coverage</span>(align_neg, <span class="at">weight =</span> 1L))</span>
<span id="cb62-25"><a href="#cb62-25" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb62-26"><a href="#cb62-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"pos"</span>, <span class="st">"neg"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb62-27"><a href="#cb62-27" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">set_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb62-28"><a href="#cb62-28" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">map</span>(<span class="cf">function</span>(strand) {</span>
<span id="cb62-29"><a href="#cb62-29" aria-hidden="true" tabindex="-1"></a>          mtx <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="dv">0</span>, <span class="at">ncol=</span>w, <span class="at">nrow=</span><span class="fu">length</span>(motif_matches))</span>
<span id="cb62-30"><a href="#cb62-30" aria-hidden="true" tabindex="-1"></a>          <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(motif_matches)) {</span>
<span id="cb62-31"><a href="#cb62-31" aria-hidden="true" tabindex="-1"></a>            chr <span class="ot">&lt;-</span> <span class="fu">as.character</span>(motif_matches[i]<span class="sc">@</span>seqnames)</span>
<span id="cb62-32"><a href="#cb62-32" aria-hidden="true" tabindex="-1"></a>            position_index <span class="ot">&lt;-</span> motif_matches[i]<span class="sc">@</span>ranges<span class="sc">@</span>start<span class="sc">:</span>(motif_matches[i]<span class="sc">@</span>ranges<span class="sc">@</span>start <span class="sc">+</span> w <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb62-33"><a href="#cb62-33" aria-hidden="true" tabindex="-1"></a>            mtx[i, ] <span class="ot">=</span> <span class="fu">as.numeric</span>(cov_list[[strand]][[chr]][position_index])</span>
<span id="cb62-34"><a href="#cb62-34" aria-hidden="true" tabindex="-1"></a>          }</span>
<span id="cb62-35"><a href="#cb62-35" aria-hidden="true" tabindex="-1"></a>          mtx</span>
<span id="cb62-36"><a href="#cb62-36" aria-hidden="true" tabindex="-1"></a>      })</span>
<span id="cb62-37"><a href="#cb62-37" aria-hidden="true" tabindex="-1"></a>  })</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plotting the means per motif as in Figure 1c.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rbind</span>(</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">pos =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(motif_counts<span class="sc">$</span>Oct4<span class="sc">$</span>pos),</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">counts_pos =</span> <span class="fu">colMeans</span>(motif_counts<span class="sc">$</span>Oct4<span class="sc">$</span>pos),</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">counts_neg =</span> <span class="sc">-</span><span class="fu">colMeans</span>(motif_counts<span class="sc">$</span>Oct4<span class="sc">$</span>neg),</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">TF =</span> <span class="st">"Oct4"</span>),</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">pos =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(motif_counts<span class="sc">$</span>Sox2<span class="sc">$</span>pos),</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">counts_pos =</span> <span class="fu">colMeans</span>(motif_counts<span class="sc">$</span>Sox2<span class="sc">$</span>pos),</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">counts_neg =</span> <span class="sc">-</span><span class="fu">colMeans</span>(motif_counts<span class="sc">$</span>Sox2<span class="sc">$</span>neg),</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">TF =</span> <span class="st">"Sox2"</span>)</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>pos, <span class="at">y=</span>counts_pos, <span class="at">color=</span>TF), <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>pos, <span class="at">y=</span>counts_neg, <span class="at">color=</span>TF), <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"Oct4"</span> <span class="ot">=</span> <span class="st">"darkred"</span>, <span class="st">"Sox2"</span> <span class="ot">=</span> <span class="st">"blue"</span>)) <span class="sc">+</span></span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Position [bp]"</span>, <span class="at">y=</span><span class="st">"Coverage"</span>, <span class="at">color=</span><span class="st">"TF"</span>) <span class="sc">+</span></span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_theme_linedraw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Plot the coverage per motif and per sequence as in Figure 1b</p>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-39_c6939a0b4a038ee9d288f727183df753">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> </span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">imap_dfr</span>(motif_counts, <span class="cf">function</span>(cov, tf) {</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">imap_dfr</span>(cov, <span class="cf">function</span>(mtx, strand) {</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>      purrr<span class="sc">::</span><span class="fu">map_dfr</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(mtx), <span class="cf">function</span>(i) {</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>        tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">seq =</span> i,</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">idx =</span> (<span class="sc">-</span>w<span class="sc">/</span><span class="dv">2</span>)<span class="sc">:</span>(w<span class="sc">/</span><span class="dv">2</span> <span class="sc">-</span><span class="dv">1</span>),</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">counts =</span> mtx[i, ],</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>                       <span class="at">s =</span> strand,</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>                       <span class="at">TF =</span> tf)</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>      })</span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a>  })</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plot via log2 FC at each position</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>pcount <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>df <span class="sc">%&gt;%</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from=</span>s, <span class="at">values_from=</span>counts) <span class="sc">%&gt;%</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(TF<span class="sc">==</span><span class="st">"Oct4"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">diff =</span> <span class="fu">log2</span>(pos<span class="sc">+</span>pcount) <span class="sc">-</span> <span class="fu">log2</span>(neg<span class="sc">+</span>pcount)) <span class="sc">%&gt;%</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>seq, <span class="at">x=</span>idx)) <span class="sc">+</span></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="fu">aes</span>(<span class="at">fill=</span>diff)) <span class="sc">+</span></span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid=</span><span class="st">"black"</span>, <span class="at">high=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Relative Postion [bp]"</span>, <span class="at">y =</span> <span class="st">"Top Oct4-Sox2 Motifs"</span>, <span class="at">fill =</span> <span class="st">"Log2 FC"</span>,</span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="st">"Oct4"</span>) <span class="sc">+</span></span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_mode</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid" width="384"></p>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">%&gt;%</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from=</span>s, <span class="at">values_from=</span>counts) <span class="sc">%&gt;%</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(TF<span class="sc">==</span><span class="st">"Sox2"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">diff =</span> <span class="fu">log2</span>(pos<span class="sc">+</span>pcount) <span class="sc">-</span> <span class="fu">log2</span>(neg<span class="sc">+</span>pcount)) <span class="sc">%&gt;%</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>seq, <span class="at">x=</span>idx)) <span class="sc">+</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="fu">aes</span>(<span class="at">fill=</span>diff)) <span class="sc">+</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid=</span><span class="st">"black"</span>, <span class="at">high=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Relative Postion [bp]"</span>, <span class="at">y =</span> <span class="st">"Top Oct4-Sox2 Motifs"</span>, <span class="at">fill =</span> <span class="st">"Log2 FC"</span>,</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="st">"Sox2"</span>) <span class="sc">+</span></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_mode</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-40-2.png" class="img-fluid" width="384"></p>
</div>
</div>
<p>Plot via facetting</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>tf <span class="ot">=</span> <span class="st">"Oct4"</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(TF<span class="sc">==</span>tf) <span class="sc">%&gt;%</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counts =</span> <span class="fu">ifelse</span>(s<span class="sc">==</span><span class="st">"pos"</span>, counts, <span class="sc">-</span>counts)) <span class="sc">%&gt;%</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(counts)</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>df <span class="sc">%&gt;%</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(TF<span class="sc">==</span>tf) <span class="sc">%&gt;%</span></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counts =</span> <span class="fu">ifelse</span>(s<span class="sc">==</span><span class="st">"pos"</span>, counts, <span class="sc">-</span>counts)) <span class="sc">%&gt;%</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>seq, <span class="at">x=</span>idx)) <span class="sc">+</span></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="fu">aes</span>(<span class="at">fill=</span>counts)) <span class="sc">+</span></span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid=</span><span class="st">"black"</span>, <span class="at">high=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">colours =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"black"</span>, <span class="st">"blue"</span>),</span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a>                       <span class="at">values =</span> scales<span class="sc">::</span><span class="fu">rescale</span>(<span class="fu">c</span>(<span class="fu">max</span>(tmp), <span class="dv">5</span>, <span class="dv">0</span>, <span class="sc">-</span><span class="dv">5</span>, <span class="fu">min</span>(tmp)))) <span class="sc">+</span></span>
<span id="cb67-14"><a href="#cb67-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb67-15"><a href="#cb67-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Relative Postion [bp]"</span>, <span class="at">y =</span> <span class="st">"Top Oct4-Sox2 Motifs"</span>, <span class="at">fill =</span> <span class="st">"Log2 FC"</span>,</span>
<span id="cb67-16"><a href="#cb67-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span>tf) <span class="sc">+</span></span>
<span id="cb67-17"><a href="#cb67-17" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_mode</span>() <span class="sc">+</span></span>
<span id="cb67-18"><a href="#cb67-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>s, <span class="at">ncol=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>tf <span class="ot">=</span> <span class="st">"Sox2"</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(TF<span class="sc">==</span>tf) <span class="sc">%&gt;%</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counts =</span> <span class="fu">ifelse</span>(s<span class="sc">==</span><span class="st">"pos"</span>, counts, <span class="sc">-</span>counts)) <span class="sc">%&gt;%</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(counts)</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>df <span class="sc">%&gt;%</span></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(TF<span class="sc">==</span>tf) <span class="sc">%&gt;%</span></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counts =</span> <span class="fu">ifelse</span>(s<span class="sc">==</span><span class="st">"pos"</span>, counts, <span class="sc">-</span>counts)) <span class="sc">%&gt;%</span></span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>seq, <span class="at">x=</span>idx)) <span class="sc">+</span></span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>(<span class="fu">aes</span>(<span class="at">fill=</span>counts)) <span class="sc">+</span></span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid=</span><span class="st">"black"</span>, <span class="at">high=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">colours =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"black"</span>, <span class="st">"blue"</span>),</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a>                       <span class="at">values =</span> scales<span class="sc">::</span><span class="fu">rescale</span>(<span class="fu">c</span>(<span class="fu">max</span>(tmp), <span class="dv">5</span>, <span class="dv">0</span>, <span class="sc">-</span><span class="dv">5</span>, <span class="fu">min</span>(tmp)))) <span class="sc">+</span></span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb69-15"><a href="#cb69-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Relative Postion [bp]"</span>, <span class="at">y =</span> <span class="st">"Top Oct4-Sox2 Motifs"</span>, <span class="at">fill =</span> <span class="st">"Log2 FC"</span>,</span>
<span id="cb69-16"><a href="#cb69-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span>tf) <span class="sc">+</span></span>
<span id="cb69-17"><a href="#cb69-17" aria-hidden="true" tabindex="-1"></a>  ggdark<span class="sc">::</span><span class="fu">dark_mode</span>() <span class="sc">+</span></span>
<span id="cb69-18"><a href="#cb69-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>s, <span class="at">ncol=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for 'fill' is already present. Adding another scale for 'fill', which
will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="01_prepare_input_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="appendix" class="level1" data-number="11">
<h1 data-number="11"><span class="header-section-number">11</span> Appendix</h1>
<section id="save-image" class="level2" data-number="11.1">
<h2 data-number="11.1" class="anchored" data-anchor-id="save-image"><span class="header-section-number">11.1</span> Save Image</h2>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-43_7ad40a31160751f2f1dc36463fb7dc16">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save.image</span>(<span class="fu">paste0</span>(output_dir, <span class="st">"tmp.RData"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="session-info" class="level2" data-number="11.2">
<h2 data-number="11.2" class="anchored" data-anchor-id="session-info"><span class="header-section-number">11.2</span> Session Info</h2>
<div class="cell" data-hash="01_prepare_input_cache/html/unnamed-chunk-44_cc9ed75eea43b0252e93b72517101bb4">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.1 (2022-06-23)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
[1] C

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] TFBSTools_1.34.0                   JASPAR2020_0.99.10                
 [3] motifmatchr_1.18.0                 furrr_0.3.0                       
 [5] future_1.27.0                      forcats_0.5.1                     
 [7] stringr_1.4.0                      dplyr_1.0.9                       
 [9] purrr_0.3.4                        readr_2.1.2                       
[11] tidyr_1.2.0                        tibble_3.1.8                      
[13] ggplot2_3.3.6                      tidyverse_1.3.2                   
[15] BSgenome.Mmusculus.UCSC.mm10_1.4.3 BSgenome_1.64.0                   
[17] BRGenomics_1.8.0                   GenomicAlignments_1.32.1          
[19] Rsamtools_2.12.0                   Biostrings_2.64.0                 
[21] XVector_0.36.0                     SummarizedExperiment_1.26.1       
[23] Biobase_2.56.0                     MatrixGenerics_1.8.1              
[25] matrixStats_0.62.0                 rtracklayer_1.56.1                
[27] GenomicRanges_1.48.0               GenomeInfoDb_1.32.3               
[29] IRanges_2.30.0                     S4Vectors_0.34.0                  
[31] BiocGenerics_0.42.0                reticulate_1.26                   

loaded via a namespace (and not attached):
  [1] readxl_1.4.0                backports_1.4.1            
  [3] plyr_1.8.7                  splines_4.2.1              
  [5] BiocParallel_1.30.3         listenv_0.8.0              
  [7] digest_0.6.29               htmltools_0.5.3            
  [9] GO.db_3.15.0                fansi_1.0.3                
 [11] magrittr_2.0.3              memoise_2.0.1              
 [13] googlesheets4_1.0.0         tzdb_0.3.0                 
 [15] globals_0.16.0              annotate_1.74.0            
 [17] modelr_0.1.8                R.utils_2.12.0             
 [19] colorspace_2.0-3            blob_1.2.3                 
 [21] rvest_1.0.2                 haven_2.5.0                
 [23] xfun_0.32                   crayon_1.5.1               
 [25] RCurl_1.98-1.8              jsonlite_1.8.0             
 [27] genefilter_1.78.0           TFMPvalue_0.0.8            
 [29] survival_3.4-0              glue_1.6.2                 
 [31] gtable_0.3.0                gargle_1.2.0               
 [33] zlibbioc_1.42.0             DelayedArray_0.22.0        
 [35] ggdark_0.2.1                plyranges_1.16.0           
 [37] scales_1.2.0                DBI_1.1.3                  
 [39] Rcpp_1.0.9                  xtable_1.8-4               
 [41] bit_4.0.4                   httr_1.4.3                 
 [43] RColorBrewer_1.1-3          ellipsis_0.3.2             
 [45] pkgconfig_2.0.3             XML_3.99-0.10              
 [47] R.methodsS3_1.8.2           farver_2.1.1               
 [49] dbplyr_2.2.1                locfit_1.5-9.6             
 [51] utf8_1.2.2                  tidyselect_1.1.2           
 [53] labeling_0.4.2              rlang_1.0.4                
 [55] reshape2_1.4.4              AnnotationDbi_1.58.0       
 [57] munsell_0.5.0               cellranger_1.1.0           
 [59] tools_4.2.1                 cachem_1.0.6               
 [61] cli_3.3.0                   DirichletMultinomial_1.38.0
 [63] generics_0.1.3              RSQLite_2.2.16             
 [65] broom_1.0.0                 evaluate_0.16              
 [67] fastmap_1.1.0               yaml_2.3.5                 
 [69] knitr_1.39                  bit64_4.0.5                
 [71] fs_1.5.2                    caTools_1.18.2             
 [73] KEGGREST_1.36.3             R.oo_1.25.0                
 [75] poweRlaw_0.70.6             pracma_2.3.8               
 [77] xml2_1.3.3                  compiler_4.2.1             
 [79] png_0.1-7                   reprex_2.0.1               
 [81] geneplotter_1.74.0          stringi_1.7.8              
 [83] lattice_0.20-45             CNEr_1.32.0                
 [85] Matrix_1.5-1                vctrs_0.4.1                
 [87] pillar_1.8.0                lifecycle_1.0.1            
 [89] bitops_1.0-7                R6_2.5.1                   
 [91] BiocIO_1.6.0                parallelly_1.32.1          
 [93] codetools_0.2-18            gtools_3.9.3               
 [95] assertthat_0.2.1            seqLogo_1.62.0             
 [97] DESeq2_1.36.0               rjson_0.2.21               
 [99] withr_2.5.0                 GenomeInfoDbData_1.2.8     
[101] parallel_4.2.1              hms_1.1.1                  
[103] grid_4.2.1                  rmarkdown_2.14             
[105] googledrive_2.0.0           lubridate_1.8.0            
[107] restfulr_0.0.15            </code></pre>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>