<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Interpretability â€“ PSLS @ BPNet</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark-d166b450ba5a8e9f7a0ab969bf6592c1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-4dc3fba39d20cba924e5db3bb4a81293.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">PSLS @ BPNet</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../code/00_introduction.html"> 
<span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../code/01_download.html"> 
<span class="menu-text">Download</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../code/02_prepare_input.html"> 
<span class="menu-text">Processing</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../code/03_APIs.html"> 
<span class="menu-text">APIs</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../code/04_training.html"> 
<span class="menu-text">Training</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../code/05_architecture_comparison.html"> 
<span class="menu-text">Architectures</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../code/06_interpretability.html" aria-current="page"> 
<span class="menu-text">Interpretability</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../code/07_mutagenesis.html"> 
<span class="menu-text">Mutagenesis</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/psl-schaefer/BPNet"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://x.com/psl_schaefer"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/philipp-sl-schaefer/"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#description" id="toc-description" class="nav-link active" data-scroll-target="#description"><span class="header-section-number">1</span> Description</a></li>
  <li><a href="#libraries" id="toc-libraries" class="nav-link" data-scroll-target="#libraries"><span class="header-section-number">2</span> Libraries</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data"><span class="header-section-number">3</span> Data</a></li>
  <li><a href="#deeplift" id="toc-deeplift" class="nav-link" data-scroll-target="#deeplift"><span class="header-section-number">4</span> DeepLift</a>
  <ul class="collapse">
  <li><a href="#klf4-e2-enhancer-supplementary-fig.2" id="toc-klf4-e2-enhancer-supplementary-fig.2" class="nav-link" data-scroll-target="#klf4-e2-enhancer-supplementary-fig.2"><span class="header-section-number">4.0.1</span> Klf4 E2 enhancer (Supplementary Fig.2)</a></li>
  <li><a href="#nanog-enhancer-supplementary-fig.2" id="toc-nanog-enhancer-supplementary-fig.2" class="nav-link" data-scroll-target="#nanog-enhancer-supplementary-fig.2"><span class="header-section-number">4.0.2</span> Nanog enhancer (Supplementary Fig.2)</a></li>
  <li><a href="#fbx15-enhancer-supplementary-fig.2" id="toc-fbx15-enhancer-supplementary-fig.2" class="nav-link" data-scroll-target="#fbx15-enhancer-supplementary-fig.2"><span class="header-section-number">4.0.3</span> Fbx15 enhancer (Supplementary Fig.2)</a></li>
  </ul></li>
  <li><a href="#input-x-gradient" id="toc-input-x-gradient" class="nav-link" data-scroll-target="#input-x-gradient"><span class="header-section-number">5</span> Input x Gradient</a>
  <ul class="collapse">
  <li><a href="#klf4-e2-enhancer-supplementary-fig.2-1" id="toc-klf4-e2-enhancer-supplementary-fig.2-1" class="nav-link" data-scroll-target="#klf4-e2-enhancer-supplementary-fig.2-1"><span class="header-section-number">5.0.1</span> Klf4 E2 enhancer (Supplementary Fig.2)</a></li>
  <li><a href="#nanog-enhancer-supplementary-fig.2-1" id="toc-nanog-enhancer-supplementary-fig.2-1" class="nav-link" data-scroll-target="#nanog-enhancer-supplementary-fig.2-1"><span class="header-section-number">5.0.2</span> Nanog enhancer (Supplementary Fig.2)</a></li>
  <li><a href="#fbx15-enhancer-supplementary-fig.2-1" id="toc-fbx15-enhancer-supplementary-fig.2-1" class="nav-link" data-scroll-target="#fbx15-enhancer-supplementary-fig.2-1"><span class="header-section-number">5.0.3</span> Fbx15 enhancer (Supplementary Fig.2)</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Interpretability</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="description" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Description</h1>
<ul>
<li>Here we use Interpretable ML tools to make sense of the model predictions.</li>
</ul>
</section>
<section id="libraries" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Libraries</h1>
<div id="c082e442" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch.nn <span class="im">as</span> nn</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch.nn.functional <span class="im">as</span> F</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch.optim <span class="im">as</span> optim</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> torch.utils.data <span class="im">import</span> Dataset, DataLoader</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> captum.attr <span class="im">import</span> DeepLift</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> modisco.visualization</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> modisco.visualization <span class="im">import</span> viz_sequence</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.config <span class="im">import</span> conf_dict</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.architectures <span class="im">import</span> <span class="op">*</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.utils <span class="im">import</span> <span class="op">*</span> </span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.loss <span class="im">import</span> <span class="op">*</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.metrics <span class="im">import</span> <span class="op">*</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.DeepLiftUtils <span class="im">import</span> <span class="op">*</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="co"># dark mode doesn't quite work for the attribution plots</span></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.style.use('dark_background')</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm</code></pre>
</div>
</div>
</section>
<section id="data" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Data</h1>
<div id="215a8b11" class="cell" data-execution_count="2">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["tf_list"] = ["Nanog"]</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["batch_size"] = 248</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["max_epochs"] = 25</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["early_stop_patience"] = 4</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["restore_best_weights"] = True</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>PRC_DIR <span class="op">=</span> Path(<span class="st">".."</span>) <span class="op">/</span> <span class="st">"prc"</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>MODELS_DIR <span class="op">=</span> Path(<span class="st">".."</span>) <span class="op">/</span> <span class="st">"trained_models"</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>FIG_DIR <span class="op">=</span> Path(<span class="st">".."</span>) <span class="op">/</span> <span class="st">"figures"</span> <span class="op">/</span> <span class="st">"interpretability"</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>FIG_DIR.mkdir(exist_ok<span class="op">=</span><span class="va">True</span>, parents<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="ee085a4b" class="cell" data-execution_count="3">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> torch.cuda.is_available():</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  device <span class="op">=</span> <span class="st">"cuda"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> torch.backends.mps.is_available():</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  device <span class="op">=</span> torch.device(<span class="st">"mps"</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  decive <span class="op">=</span> <span class="st">"cpu"</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Using </span><span class="sc">{</span>device<span class="sc">}</span><span class="ss"> device"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Using mps device</code></pre>
</div>
</div>
<div id="9a47506d" class="cell" data-execution_count="4">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>train_dataset <span class="op">=</span> ChIP_Nexus_Dataset(set_name<span class="op">=</span><span class="st">"train"</span>, </span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                                   input_dir<span class="op">=</span>PRC_DIR, </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                                   TF_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>])</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>train_dataset</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="4">
<pre><code>ChIP_Nexus_Dataset
Set: train
TFs: ['Nanog', 'Klf4', 'Oct4', 'Sox2']
Size: 93904</code></pre>
</div>
</div>
<div id="2f816508" class="cell" data-execution_count="5">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>tune_dataset <span class="op">=</span> ChIP_Nexus_Dataset(set_name<span class="op">=</span><span class="st">"tune"</span>, </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                                  input_dir<span class="op">=</span>PRC_DIR, </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                                  TF_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>])</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>tune_dataset</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>ChIP_Nexus_Dataset
Set: tune
TFs: ['Nanog', 'Klf4', 'Oct4', 'Sox2']
Size: 29277</code></pre>
</div>
</div>
<div id="cf89dfff" class="cell" data-execution_count="6">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>test_dataset <span class="op">=</span> ChIP_Nexus_Dataset(set_name<span class="op">=</span><span class="st">"test"</span>, </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                                  input_dir<span class="op">=</span>PRC_DIR, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                                  TF_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>])</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>test_dataset</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>ChIP_Nexus_Dataset
Set: test
TFs: ['Nanog', 'Klf4', 'Oct4', 'Sox2']
Size: 27727</code></pre>
</div>
</div>
<p>For interpretability methods like DeepLift (https://github.com/kundajelab/deeplift), integrated gradients or gradient x input, we require a scalar output. The BPNet model does, however, predict profiles shapes as tensors of size 2 x 1000 (strand x bps). For backpropagation or DeepLift we collapse the profile for each strand to one representative value, which can then be used to compute the gradient of this scalar with respect to each of the input bps.</p>
<p>We define our profile prediction as the softmax of the pre-activation of the last layer in our profile shape prediction head.</p>
<p><span class="math inline">\(\tilde{z}\)</span> is the pre-activation of our last layer.</p>
<p><span class="math inline">\(p = softmax(\tilde{z}) = \frac{\exp{\tilde{z}}}{\sum_{i}^{N}{\exp{\tilde{z'}_i}}}\)</span> with <span class="math inline">\(N\)</span> corresponding to the number of bps.</p>
<p>We weight the pre-activations of the last layer with the softmax of the same pre-activations and take the sum. This way we get one value for each strand.</p>
<p><span class="math inline">\(z = \sum_{i}^{N}{p_i * \tilde{z}_i}\)</span> with <span class="math inline">\(z \in \mathbb{R}^{2}\)</span></p>
<p>Importantly, we have to detach the softmax activation so that it is a constant value during backpropagation of contribution scores or gradients.</p>
<p>When computing DeepLift scores we take the average of the two strands.</p>
<div id="e2cb3737" class="cell" data-execution_count="7">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> torch.load(MODELS_DIR <span class="op">/</span> <span class="st">"all_tfs_model.pt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/var/folders/n8/xnf8qmpj41795tf723xy8mqm0000gn/T/ipykernel_94928/1827116589.py:1: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  model = torch.load(MODELS_DIR / "all_tfs_model.pt")</code></pre>
</div>
</div>
</section>
<section id="deeplift" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> DeepLift</h1>
<p>We picked some exemplary regions of interest from the paper (Fig.2 and Supp.Fig.2) to compute the contribution scores and make some plots.</p>
<div id="370661dd" class="cell" data-execution_count="8">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>dl <span class="op">=</span> DeepLift(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="117be393" class="cell" data-execution_count="9">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_contr_region(seqname, start, end, dataset, model, dl, device, tf_list, output_dir, plot<span class="op">=</span><span class="va">True</span>, figsize1<span class="op">=</span>(<span class="dv">20</span>,<span class="dv">2</span>), figsize2<span class="op">=</span>(<span class="dv">10</span>,<span class="fl">1.5</span>)):</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Compute the DeepLift contribution scores for a 1kb sequence which contains the shorter </span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co">    region of interest specified by the input arguments.</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co">    Params:</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co">        seq_name: string</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co">            specifies the chromosome </span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co">        start: int</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co">            specifies start coordinate of sequence of interest on chromosome</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="co">        end: int</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co">            specifies end coordinate of sequence of interest on chromsome</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co">        dataset: utils.ChIP_Nexus_Dataset object</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co">        device: cuda or cpu</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="co">        tf_list: </span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="co">            Contains names of TFs for which we want to compute the contributions</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="co">        plot: bool</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="co">            Whether to visualize the DeepLift contribution scores.</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a><span class="co">        contr: tensor (4x1000)</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="co">            Contains the contribution of each bp to the profile shape predictions for the input sequence.</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="co">        dist_start: int</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="co">            distance between the start of the 1kb sequence and the region of interest</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># select sequence of interest</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>    tmp_df, idx, dist_start, one_hot, baseline, bias_raw, bias_smooth, tf_counts <span class="op">=</span> get_seq_oi(seqname, start, end, dataset, device)</span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>    width <span class="op">=</span> end <span class="op">-</span> start</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>    <span class="co"># compute contribution scores for each tf</span></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>    contr_list <span class="op">=</span> []</span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>    df_list <span class="op">=</span> []</span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> tf_index, tf <span class="kw">in</span> <span class="bu">enumerate</span>(tf_list):</span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>        contr <span class="op">=</span> dl.attribute(inputs<span class="op">=</span>one_hot, baselines<span class="op">=</span>baseline, target<span class="op">=</span>(tf_index), additional_forward_args<span class="op">=</span>(bias_raw, bias_smooth, <span class="va">True</span>)).detach().cpu().numpy()</span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>        contr_list.append(contr)</span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>        pred, _ <span class="op">=</span> model.forward(one_hot, bias_raw, bias_smooth, interpretation<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>        pred <span class="op">=</span> pred.detach().cpu().numpy().squeeze()</span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a>        <span class="co"># scale prediciton with total counts</span></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>        pred <span class="op">=</span> pred <span class="op">*</span> tf_counts.<span class="bu">sum</span>(axis<span class="op">=-</span><span class="dv">1</span>, keepdims<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>        tf_df <span class="op">=</span> pd.DataFrame({<span class="st">"pos"</span>: np.arange(width<span class="op">+</span><span class="dv">1</span>), <span class="st">"TF"</span>: tf, <span class="st">"pos_values"</span>: pred[tf_index, <span class="dv">0</span>, dist_start : (dist_start <span class="op">+</span> width<span class="op">+</span><span class="dv">1</span>)], <span class="st">"neg_values"</span>: pred[tf_index, <span class="dv">1</span>, dist_start : (dist_start <span class="op">+</span> width<span class="op">+</span><span class="dv">1</span>)]})</span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>        df_list.append(tf_df)</span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> plot:</span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a>            <span class="co"># entire sequence original</span></span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a>            plot_weights(contr,</span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a>            fontsizes<span class="op">=</span>[<span class="dv">20</span>,<span class="dv">15</span>,<span class="dv">15</span>],</span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a>            title <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>tf<span class="sc">}</span><span class="ss"> - 1kbp sequence"</span>, </span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a>            xlabel<span class="op">=</span><span class="ss">f"</span><span class="sc">{</span>tmp_df<span class="sc">.</span>seqnames[idx]<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span>tmp_df<span class="sc">.</span>start[idx]<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>tmp_df<span class="sc">.</span>end[idx]<span class="sc">}</span><span class="ss">"</span>, </span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a>            ylabel<span class="op">=</span><span class="st">"DeepLift contribution scores"</span>,</span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a>            subticks_frequency<span class="op">=</span><span class="dv">20</span>, figsize<span class="op">=</span>figsize1)</span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a>            plt.savefig(output_dir <span class="op">/</span> <span class="ss">f"</span><span class="sc">{</span>tf<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>seqname<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>start<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>end<span class="sc">}</span><span class="ss">_entireSeq_DeepLift.pdf"</span>)</span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a>            <span class="co"># zoomed into motif region</span></span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a>            plot_weights(contr[:, :,dist_start : (dist_start <span class="op">+</span> width<span class="op">+</span><span class="dv">1</span>)],</span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a>            fontsizes<span class="op">=</span>[<span class="dv">20</span>,<span class="dv">15</span>,<span class="dv">15</span>],</span>
<span id="cb15-55"><a href="#cb15-55" aria-hidden="true" tabindex="-1"></a>            title <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>tf<span class="sc">}</span><span class="ss"> - Motif of interest"</span>, </span>
<span id="cb15-56"><a href="#cb15-56" aria-hidden="true" tabindex="-1"></a>            xlabel<span class="op">=</span><span class="ss">f"</span><span class="sc">{</span>seqname<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span>start<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>end<span class="sc">}</span><span class="ss">, (</span><span class="sc">{</span>dist_start<span class="sc">}</span><span class="ss"> - </span><span class="sc">{</span>dist_start <span class="op">+</span> width<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">)"</span>, </span>
<span id="cb15-57"><a href="#cb15-57" aria-hidden="true" tabindex="-1"></a>            ylabel<span class="op">=</span><span class="st">"DeepLift contribution scores"</span>,</span>
<span id="cb15-58"><a href="#cb15-58" aria-hidden="true" tabindex="-1"></a>            subticks_frequency<span class="op">=</span><span class="dv">10</span>, figsize<span class="op">=</span>figsize2)</span>
<span id="cb15-59"><a href="#cb15-59" aria-hidden="true" tabindex="-1"></a>            plt.savefig(output_dir <span class="op">/</span> <span class="ss">f"</span><span class="sc">{</span>tf<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>seqname<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>start<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>end<span class="sc">}</span><span class="ss">_zoomedSeq_DeepLift.pdf"</span>)  </span>
<span id="cb15-60"><a href="#cb15-60" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb15-61"><a href="#cb15-61" aria-hidden="true" tabindex="-1"></a>            <span class="co"># plot profiles</span></span>
<span id="cb15-62"><a href="#cb15-62" aria-hidden="true" tabindex="-1"></a>            fig, axis <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">2</span>,figsize<span class="op">=</span>(<span class="dv">12</span>,<span class="dv">4</span>))</span>
<span id="cb15-63"><a href="#cb15-63" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].plot(tf_counts[tf_index, <span class="dv">0</span>, :], label<span class="op">=</span><span class="st">"true counts"</span>, color<span class="op">=</span><span class="st">"green"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb15-64"><a href="#cb15-64" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].plot(<span class="op">-</span>tf_counts[tf_index, <span class="dv">1</span>, :], color<span class="op">=</span><span class="st">"green"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb15-65"><a href="#cb15-65" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].plot(pred[tf_index, <span class="dv">0</span>, :], label<span class="op">=</span><span class="st">"pred"</span>, color<span class="op">=</span><span class="st">"blue"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb15-66"><a href="#cb15-66" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].plot(<span class="op">-</span>pred[tf_index, <span class="dv">1</span>, :], color<span class="op">=</span><span class="st">"blue"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)   </span>
<span id="cb15-67"><a href="#cb15-67" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].set_xlabel(<span class="st">"bp"</span>)</span>
<span id="cb15-68"><a href="#cb15-68" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].set_ylabel(<span class="st">"Read counts"</span>)</span>
<span id="cb15-69"><a href="#cb15-69" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].plot(pred[tf_index, <span class="dv">0</span>, :], label<span class="op">=</span><span class="st">"pred"</span>, color<span class="op">=</span><span class="st">"blue"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb15-70"><a href="#cb15-70" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].plot(<span class="op">-</span>pred[tf_index, <span class="dv">1</span>, :], color<span class="op">=</span><span class="st">"blue"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb15-71"><a href="#cb15-71" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].set_xlabel(<span class="st">"bp"</span>)</span>
<span id="cb15-72"><a href="#cb15-72" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].set_ylabel(<span class="st">"Predicted probabilitiy * total counts"</span>)</span>
<span id="cb15-73"><a href="#cb15-73" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].legend()</span>
<span id="cb15-74"><a href="#cb15-74" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].legend()</span>
<span id="cb15-75"><a href="#cb15-75" aria-hidden="true" tabindex="-1"></a>            plt.show()</span>
<span id="cb15-76"><a href="#cb15-76" aria-hidden="true" tabindex="-1"></a>    plot_df <span class="op">=</span> pd.concat(df_list)</span>
<span id="cb15-77"><a href="#cb15-77" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb15-78"><a href="#cb15-78" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> contr, dist_start, plot_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="klf4-e2-enhancer-supplementary-fig.2" class="level3" data-number="4.0.1">
<h3 data-number="4.0.1" class="anchored" data-anchor-id="klf4-e2-enhancer-supplementary-fig.2"><span class="header-section-number">4.0.1</span> Klf4 E2 enhancer (Supplementary Fig.2)</h3>
<div id="c207b3de" class="cell" data-execution_count="10">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>contr, dist_start, pred <span class="op">=</span> get_contr_region(<span class="st">"chr4"</span>, start<span class="op">=</span><span class="dv">55475545</span>, end<span class="op">=</span><span class="dv">55475604</span>, dataset<span class="op">=</span>tune_dataset, output_dir<span class="op">=</span>FIG_DIR, dl<span class="op">=</span>dl, tf_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>], device<span class="op">=</span>device,   model<span class="op">=</span>model)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co">#pred.to_csv("/home/kathi/AML_Project/data/test_fig.csv")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-2.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-3.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-4.png" width="968" height="356" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-6.png" width="1598" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-7.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-8.png" width="968" height="356" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-10.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-11.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-12.png" width="972" height="356" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-14.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-15.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-11-output-16.png" width="968" height="356" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="nanog-enhancer-supplementary-fig.2" class="level3" data-number="4.0.2">
<h3 data-number="4.0.2" class="anchored" data-anchor-id="nanog-enhancer-supplementary-fig.2"><span class="header-section-number">4.0.2</span> Nanog enhancer (Supplementary Fig.2)</h3>
<div id="f5b85284" class="cell" data-execution_count="11">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>contr, dist_start, pred <span class="op">=</span> get_contr_region(<span class="st">"chr6"</span>, start<span class="op">=</span><span class="dv">122707394</span>, end<span class="op">=</span><span class="dv">122707454</span>, dataset<span class="op">=</span>train_dataset, output_dir<span class="op">=</span>FIG_DIR, dl<span class="op">=</span>dl, tf_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>], device<span class="op">=</span>device, model<span class="op">=</span>model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-2.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-3.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-4.png" width="968" height="356" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-6.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-7.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-8.png" width="968" height="356" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-10.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-11.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-12.png" width="968" height="356" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-14.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-15.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-12-output-16.png" width="959" height="356" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="fbx15-enhancer-supplementary-fig.2" class="level3" data-number="4.0.3">
<h3 data-number="4.0.3" class="anchored" data-anchor-id="fbx15-enhancer-supplementary-fig.2"><span class="header-section-number">4.0.3</span> Fbx15 enhancer (Supplementary Fig.2)</h3>
<div id="0faaed90" class="cell" data-execution_count="12">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>contr, dist_start, pred <span class="op">=</span> get_contr_region(<span class="st">"chr18"</span>, start<span class="op">=</span><span class="dv">84934461</span>, end<span class="op">=</span><span class="dv">84934521</span>, dataset<span class="op">=</span>train_dataset, output_dir<span class="op">=</span>FIG_DIR, dl<span class="op">=</span>dl, tf_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>], device<span class="op">=</span>device, model<span class="op">=</span>model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-2.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-3.png" width="819" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-4.png" width="968" height="356" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-6.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-7.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-8.png" width="968" height="356" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-10.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-11.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-12.png" width="959" height="357" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/_utils/gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
/Users/pschafer/miniforge3/envs/py-env-bpnet/lib/python3.12/site-packages/captum/attr/_core/deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-14.png" width="1590" height="243" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-15.png" width="830" height="206" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="06_interpretability_files/figure-html/cell-13-output-16.png" width="959" height="356" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="input-x-gradient" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Input x Gradient</h1>
<p>For the same exemplary regions from the paper.</p>
<p><code>#{python} grad, grad_in = input_gradient("chr4", start=55475545, end=55475604, dataset=tune_dataset,output_dir=FIG_DIR,  model=model, tf_list=conf_dict["tf_list"], device=device)</code></p>
<section id="klf4-e2-enhancer-supplementary-fig.2-1" class="level3" data-number="5.0.1">
<h3 data-number="5.0.1" class="anchored" data-anchor-id="klf4-e2-enhancer-supplementary-fig.2-1"><span class="header-section-number">5.0.1</span> Klf4 E2 enhancer (Supplementary Fig.2)</h3>
<p><code>#{python} grad, grad_in = input_gradient("chr4", start=55475545, end=55475604, dataset=tune_dataset,output_dir=FIG_DIR,   model=model, tf_list=conf_dict["tf_list"], device=device)</code></p>
</section>
<section id="nanog-enhancer-supplementary-fig.2-1" class="level3" data-number="5.0.2">
<h3 data-number="5.0.2" class="anchored" data-anchor-id="nanog-enhancer-supplementary-fig.2-1"><span class="header-section-number">5.0.2</span> Nanog enhancer (Supplementary Fig.2)</h3>
<p><code>#{python} grad, grad_in = input_gradient("chr6", start=122707394, end=122707454, dataset=train_dataset, output_dir=FIG_DIR,  model=model, tf_list=conf_dict["tf_list"], device=device)</code></p>
</section>
<section id="fbx15-enhancer-supplementary-fig.2-1" class="level3" data-number="5.0.3">
<h3 data-number="5.0.3" class="anchored" data-anchor-id="fbx15-enhancer-supplementary-fig.2-1"><span class="header-section-number">5.0.3</span> Fbx15 enhancer (Supplementary Fig.2)</h3>
<p><code>#{python} grad, grad_in = input_gradient("chr17", start=35504453, end=35504603, dataset=train_dataset, output_dir=FIG_DIR,  model=model, tf_list=conf_dict["tf_list"], device=device)</code></p>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb31" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Interpretability"</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="an">jupyter:</span><span class="co"> py-env-bpnet</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co">  freeze: auto</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="fu"># Description</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Here we use Interpretable ML tools to make sense of the model predictions.</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a><span class="fu"># Libraries</span></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch.nn <span class="im">as</span> nn</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch.nn.functional <span class="im">as</span> F</span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> torch.optim <span class="im">as</span> optim</span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> torch.utils.data <span class="im">import</span> Dataset, DataLoader</span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> captum.attr <span class="im">import</span> DeepLift</span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt</span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> modisco.visualization</span>
<span id="cb31-29"><a href="#cb31-29" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> modisco.visualization <span class="im">import</span> viz_sequence</span>
<span id="cb31-30"><a href="#cb31-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-31"><a href="#cb31-31" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.config <span class="im">import</span> conf_dict</span>
<span id="cb31-32"><a href="#cb31-32" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.architectures <span class="im">import</span> <span class="op">*</span></span>
<span id="cb31-33"><a href="#cb31-33" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.utils <span class="im">import</span> <span class="op">*</span> </span>
<span id="cb31-34"><a href="#cb31-34" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.loss <span class="im">import</span> <span class="op">*</span></span>
<span id="cb31-35"><a href="#cb31-35" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.metrics <span class="im">import</span> <span class="op">*</span></span>
<span id="cb31-36"><a href="#cb31-36" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> src.DeepLiftUtils <span class="im">import</span> <span class="op">*</span></span>
<span id="cb31-37"><a href="#cb31-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-38"><a href="#cb31-38" aria-hidden="true" tabindex="-1"></a><span class="co"># dark mode doesn't quite work for the attribution plots</span></span>
<span id="cb31-39"><a href="#cb31-39" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.style.use('dark_background')</span></span>
<span id="cb31-40"><a href="#cb31-40" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-41"><a href="#cb31-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-42"><a href="#cb31-42" aria-hidden="true" tabindex="-1"></a><span class="fu"># Data</span></span>
<span id="cb31-43"><a href="#cb31-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-46"><a href="#cb31-46" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-47"><a href="#cb31-47" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["tf_list"] = ["Nanog"]</span></span>
<span id="cb31-48"><a href="#cb31-48" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["batch_size"] = 248</span></span>
<span id="cb31-49"><a href="#cb31-49" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["max_epochs"] = 25</span></span>
<span id="cb31-50"><a href="#cb31-50" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["early_stop_patience"] = 4</span></span>
<span id="cb31-51"><a href="#cb31-51" aria-hidden="true" tabindex="-1"></a><span class="co">#conf_dict["restore_best_weights"] = True</span></span>
<span id="cb31-52"><a href="#cb31-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-53"><a href="#cb31-53" aria-hidden="true" tabindex="-1"></a>PRC_DIR <span class="op">=</span> Path(<span class="st">".."</span>) <span class="op">/</span> <span class="st">"prc"</span></span>
<span id="cb31-54"><a href="#cb31-54" aria-hidden="true" tabindex="-1"></a>MODELS_DIR <span class="op">=</span> Path(<span class="st">".."</span>) <span class="op">/</span> <span class="st">"trained_models"</span></span>
<span id="cb31-55"><a href="#cb31-55" aria-hidden="true" tabindex="-1"></a>FIG_DIR <span class="op">=</span> Path(<span class="st">".."</span>) <span class="op">/</span> <span class="st">"figures"</span> <span class="op">/</span> <span class="st">"interpretability"</span></span>
<span id="cb31-56"><a href="#cb31-56" aria-hidden="true" tabindex="-1"></a>FIG_DIR.mkdir(exist_ok<span class="op">=</span><span class="va">True</span>, parents<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb31-57"><a href="#cb31-57" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-58"><a href="#cb31-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-61"><a href="#cb31-61" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-62"><a href="#cb31-62" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> torch.cuda.is_available():</span>
<span id="cb31-63"><a href="#cb31-63" aria-hidden="true" tabindex="-1"></a>  device <span class="op">=</span> <span class="st">"cuda"</span></span>
<span id="cb31-64"><a href="#cb31-64" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> torch.backends.mps.is_available():</span>
<span id="cb31-65"><a href="#cb31-65" aria-hidden="true" tabindex="-1"></a>  device <span class="op">=</span> torch.device(<span class="st">"mps"</span>)</span>
<span id="cb31-66"><a href="#cb31-66" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb31-67"><a href="#cb31-67" aria-hidden="true" tabindex="-1"></a>  decive <span class="op">=</span> <span class="st">"cpu"</span></span>
<span id="cb31-68"><a href="#cb31-68" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Using </span><span class="sc">{</span>device<span class="sc">}</span><span class="ss"> device"</span>)</span>
<span id="cb31-69"><a href="#cb31-69" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-70"><a href="#cb31-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-73"><a href="#cb31-73" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-74"><a href="#cb31-74" aria-hidden="true" tabindex="-1"></a>train_dataset <span class="op">=</span> ChIP_Nexus_Dataset(set_name<span class="op">=</span><span class="st">"train"</span>, </span>
<span id="cb31-75"><a href="#cb31-75" aria-hidden="true" tabindex="-1"></a>                                   input_dir<span class="op">=</span>PRC_DIR, </span>
<span id="cb31-76"><a href="#cb31-76" aria-hidden="true" tabindex="-1"></a>                                   TF_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>])</span>
<span id="cb31-77"><a href="#cb31-77" aria-hidden="true" tabindex="-1"></a>train_dataset</span>
<span id="cb31-78"><a href="#cb31-78" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-79"><a href="#cb31-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-82"><a href="#cb31-82" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-83"><a href="#cb31-83" aria-hidden="true" tabindex="-1"></a>tune_dataset <span class="op">=</span> ChIP_Nexus_Dataset(set_name<span class="op">=</span><span class="st">"tune"</span>, </span>
<span id="cb31-84"><a href="#cb31-84" aria-hidden="true" tabindex="-1"></a>                                  input_dir<span class="op">=</span>PRC_DIR, </span>
<span id="cb31-85"><a href="#cb31-85" aria-hidden="true" tabindex="-1"></a>                                  TF_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>])</span>
<span id="cb31-86"><a href="#cb31-86" aria-hidden="true" tabindex="-1"></a>tune_dataset</span>
<span id="cb31-87"><a href="#cb31-87" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-88"><a href="#cb31-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-91"><a href="#cb31-91" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-92"><a href="#cb31-92" aria-hidden="true" tabindex="-1"></a>test_dataset <span class="op">=</span> ChIP_Nexus_Dataset(set_name<span class="op">=</span><span class="st">"test"</span>, </span>
<span id="cb31-93"><a href="#cb31-93" aria-hidden="true" tabindex="-1"></a>                                  input_dir<span class="op">=</span>PRC_DIR, </span>
<span id="cb31-94"><a href="#cb31-94" aria-hidden="true" tabindex="-1"></a>                                  TF_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>])</span>
<span id="cb31-95"><a href="#cb31-95" aria-hidden="true" tabindex="-1"></a>test_dataset</span>
<span id="cb31-96"><a href="#cb31-96" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-97"><a href="#cb31-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-98"><a href="#cb31-98" aria-hidden="true" tabindex="-1"></a>For interpretability methods like DeepLift (https://github.com/kundajelab/deeplift), integrated gradients or gradient x input, we require a scalar output. The BPNet model does, however, predict profiles shapes as tensors of size 2 x 1000 (strand x bps). For backpropagation or DeepLift we collapse the profile for each strand to one representative value, which can then be used to compute the gradient of this scalar with respect to each of the input bps. </span>
<span id="cb31-99"><a href="#cb31-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-100"><a href="#cb31-100" aria-hidden="true" tabindex="-1"></a>We define our profile prediction as the softmax of the pre-activation of the last layer in our profile shape prediction head.</span>
<span id="cb31-101"><a href="#cb31-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-102"><a href="#cb31-102" aria-hidden="true" tabindex="-1"></a>$\tilde{z}$ is the pre-activation of our last layer.</span>
<span id="cb31-103"><a href="#cb31-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-104"><a href="#cb31-104" aria-hidden="true" tabindex="-1"></a>$p = softmax(\tilde{z}) = \frac{\exp{\tilde{z}}}{\sum_{i}^{N}{\exp{\tilde{z'}_i}}}$ with $N$ corresponding to the number of bps.</span>
<span id="cb31-105"><a href="#cb31-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-106"><a href="#cb31-106" aria-hidden="true" tabindex="-1"></a>We weight the pre-activations of the last layer with the softmax of the same pre-activations and take the sum. This way we get one value for each strand.</span>
<span id="cb31-107"><a href="#cb31-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-108"><a href="#cb31-108" aria-hidden="true" tabindex="-1"></a>$z = \sum_{i}^{N}{p_i * \tilde{z}_i}$ with $z \in \mathbb{R}^{2}$</span>
<span id="cb31-109"><a href="#cb31-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-110"><a href="#cb31-110" aria-hidden="true" tabindex="-1"></a>Importantly, we have to detach the softmax activation so that it is a constant value during backpropagation of contribution scores or gradients.</span>
<span id="cb31-111"><a href="#cb31-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-112"><a href="#cb31-112" aria-hidden="true" tabindex="-1"></a>When computing DeepLift scores we  take the average of the two strands. </span>
<span id="cb31-113"><a href="#cb31-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-116"><a href="#cb31-116" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-117"><a href="#cb31-117" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> torch.load(MODELS_DIR <span class="op">/</span> <span class="st">"all_tfs_model.pt"</span>)</span>
<span id="cb31-118"><a href="#cb31-118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-119"><a href="#cb31-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-120"><a href="#cb31-120" aria-hidden="true" tabindex="-1"></a><span class="fu"># DeepLift</span></span>
<span id="cb31-121"><a href="#cb31-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-122"><a href="#cb31-122" aria-hidden="true" tabindex="-1"></a>We picked some exemplary regions of interest from the paper (Fig.2 and Supp.Fig.2) to compute the contribution scores and make some plots.</span>
<span id="cb31-123"><a href="#cb31-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-126"><a href="#cb31-126" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-127"><a href="#cb31-127" aria-hidden="true" tabindex="-1"></a>dl <span class="op">=</span> DeepLift(model)</span>
<span id="cb31-128"><a href="#cb31-128" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-129"><a href="#cb31-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-132"><a href="#cb31-132" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-133"><a href="#cb31-133" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_contr_region(seqname, start, end, dataset, model, dl, device, tf_list, output_dir, plot<span class="op">=</span><span class="va">True</span>, figsize1<span class="op">=</span>(<span class="dv">20</span>,<span class="dv">2</span>), figsize2<span class="op">=</span>(<span class="dv">10</span>,<span class="fl">1.5</span>)):</span>
<span id="cb31-134"><a href="#cb31-134" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Compute the DeepLift contribution scores for a 1kb sequence which contains the shorter </span></span>
<span id="cb31-135"><a href="#cb31-135" aria-hidden="true" tabindex="-1"></a><span class="co">    region of interest specified by the input arguments.</span></span>
<span id="cb31-136"><a href="#cb31-136" aria-hidden="true" tabindex="-1"></a><span class="co">    Params:</span></span>
<span id="cb31-137"><a href="#cb31-137" aria-hidden="true" tabindex="-1"></a><span class="co">        seq_name: string</span></span>
<span id="cb31-138"><a href="#cb31-138" aria-hidden="true" tabindex="-1"></a><span class="co">            specifies the chromosome </span></span>
<span id="cb31-139"><a href="#cb31-139" aria-hidden="true" tabindex="-1"></a><span class="co">        start: int</span></span>
<span id="cb31-140"><a href="#cb31-140" aria-hidden="true" tabindex="-1"></a><span class="co">            specifies start coordinate of sequence of interest on chromosome</span></span>
<span id="cb31-141"><a href="#cb31-141" aria-hidden="true" tabindex="-1"></a><span class="co">        end: int</span></span>
<span id="cb31-142"><a href="#cb31-142" aria-hidden="true" tabindex="-1"></a><span class="co">            specifies end coordinate of sequence of interest on chromsome</span></span>
<span id="cb31-143"><a href="#cb31-143" aria-hidden="true" tabindex="-1"></a><span class="co">        dataset: utils.ChIP_Nexus_Dataset object</span></span>
<span id="cb31-144"><a href="#cb31-144" aria-hidden="true" tabindex="-1"></a><span class="co">        device: cuda or cpu</span></span>
<span id="cb31-145"><a href="#cb31-145" aria-hidden="true" tabindex="-1"></a><span class="co">        tf_list: </span></span>
<span id="cb31-146"><a href="#cb31-146" aria-hidden="true" tabindex="-1"></a><span class="co">            Contains names of TFs for which we want to compute the contributions</span></span>
<span id="cb31-147"><a href="#cb31-147" aria-hidden="true" tabindex="-1"></a><span class="co">        plot: bool</span></span>
<span id="cb31-148"><a href="#cb31-148" aria-hidden="true" tabindex="-1"></a><span class="co">            Whether to visualize the DeepLift contribution scores.</span></span>
<span id="cb31-149"><a href="#cb31-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-150"><a href="#cb31-150" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb31-151"><a href="#cb31-151" aria-hidden="true" tabindex="-1"></a><span class="co">        contr: tensor (4x1000)</span></span>
<span id="cb31-152"><a href="#cb31-152" aria-hidden="true" tabindex="-1"></a><span class="co">            Contains the contribution of each bp to the profile shape predictions for the input sequence.</span></span>
<span id="cb31-153"><a href="#cb31-153" aria-hidden="true" tabindex="-1"></a><span class="co">        dist_start: int</span></span>
<span id="cb31-154"><a href="#cb31-154" aria-hidden="true" tabindex="-1"></a><span class="co">            distance between the start of the 1kb sequence and the region of interest</span></span>
<span id="cb31-155"><a href="#cb31-155" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb31-156"><a href="#cb31-156" aria-hidden="true" tabindex="-1"></a>    <span class="co"># select sequence of interest</span></span>
<span id="cb31-157"><a href="#cb31-157" aria-hidden="true" tabindex="-1"></a>    tmp_df, idx, dist_start, one_hot, baseline, bias_raw, bias_smooth, tf_counts <span class="op">=</span> get_seq_oi(seqname, start, end, dataset, device)</span>
<span id="cb31-158"><a href="#cb31-158" aria-hidden="true" tabindex="-1"></a>    width <span class="op">=</span> end <span class="op">-</span> start</span>
<span id="cb31-159"><a href="#cb31-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-160"><a href="#cb31-160" aria-hidden="true" tabindex="-1"></a>    <span class="co"># compute contribution scores for each tf</span></span>
<span id="cb31-161"><a href="#cb31-161" aria-hidden="true" tabindex="-1"></a>    contr_list <span class="op">=</span> []</span>
<span id="cb31-162"><a href="#cb31-162" aria-hidden="true" tabindex="-1"></a>    df_list <span class="op">=</span> []</span>
<span id="cb31-163"><a href="#cb31-163" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> tf_index, tf <span class="kw">in</span> <span class="bu">enumerate</span>(tf_list):</span>
<span id="cb31-164"><a href="#cb31-164" aria-hidden="true" tabindex="-1"></a>        contr <span class="op">=</span> dl.attribute(inputs<span class="op">=</span>one_hot, baselines<span class="op">=</span>baseline, target<span class="op">=</span>(tf_index), additional_forward_args<span class="op">=</span>(bias_raw, bias_smooth, <span class="va">True</span>)).detach().cpu().numpy()</span>
<span id="cb31-165"><a href="#cb31-165" aria-hidden="true" tabindex="-1"></a>        contr_list.append(contr)</span>
<span id="cb31-166"><a href="#cb31-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-167"><a href="#cb31-167" aria-hidden="true" tabindex="-1"></a>        pred, _ <span class="op">=</span> model.forward(one_hot, bias_raw, bias_smooth, interpretation<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb31-168"><a href="#cb31-168" aria-hidden="true" tabindex="-1"></a>        pred <span class="op">=</span> pred.detach().cpu().numpy().squeeze()</span>
<span id="cb31-169"><a href="#cb31-169" aria-hidden="true" tabindex="-1"></a>        <span class="co"># scale prediciton with total counts</span></span>
<span id="cb31-170"><a href="#cb31-170" aria-hidden="true" tabindex="-1"></a>        pred <span class="op">=</span> pred <span class="op">*</span> tf_counts.<span class="bu">sum</span>(axis<span class="op">=-</span><span class="dv">1</span>, keepdims<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb31-171"><a href="#cb31-171" aria-hidden="true" tabindex="-1"></a>        tf_df <span class="op">=</span> pd.DataFrame({<span class="st">"pos"</span>: np.arange(width<span class="op">+</span><span class="dv">1</span>), <span class="st">"TF"</span>: tf, <span class="st">"pos_values"</span>: pred[tf_index, <span class="dv">0</span>, dist_start : (dist_start <span class="op">+</span> width<span class="op">+</span><span class="dv">1</span>)], <span class="st">"neg_values"</span>: pred[tf_index, <span class="dv">1</span>, dist_start : (dist_start <span class="op">+</span> width<span class="op">+</span><span class="dv">1</span>)]})</span>
<span id="cb31-172"><a href="#cb31-172" aria-hidden="true" tabindex="-1"></a>        df_list.append(tf_df)</span>
<span id="cb31-173"><a href="#cb31-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-174"><a href="#cb31-174" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> plot:</span>
<span id="cb31-175"><a href="#cb31-175" aria-hidden="true" tabindex="-1"></a>            <span class="co"># entire sequence original</span></span>
<span id="cb31-176"><a href="#cb31-176" aria-hidden="true" tabindex="-1"></a>            plot_weights(contr,</span>
<span id="cb31-177"><a href="#cb31-177" aria-hidden="true" tabindex="-1"></a>            fontsizes<span class="op">=</span>[<span class="dv">20</span>,<span class="dv">15</span>,<span class="dv">15</span>],</span>
<span id="cb31-178"><a href="#cb31-178" aria-hidden="true" tabindex="-1"></a>            title <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>tf<span class="sc">}</span><span class="ss"> - 1kbp sequence"</span>, </span>
<span id="cb31-179"><a href="#cb31-179" aria-hidden="true" tabindex="-1"></a>            xlabel<span class="op">=</span><span class="ss">f"</span><span class="sc">{</span>tmp_df<span class="sc">.</span>seqnames[idx]<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span>tmp_df<span class="sc">.</span>start[idx]<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>tmp_df<span class="sc">.</span>end[idx]<span class="sc">}</span><span class="ss">"</span>, </span>
<span id="cb31-180"><a href="#cb31-180" aria-hidden="true" tabindex="-1"></a>            ylabel<span class="op">=</span><span class="st">"DeepLift contribution scores"</span>,</span>
<span id="cb31-181"><a href="#cb31-181" aria-hidden="true" tabindex="-1"></a>            subticks_frequency<span class="op">=</span><span class="dv">20</span>, figsize<span class="op">=</span>figsize1)</span>
<span id="cb31-182"><a href="#cb31-182" aria-hidden="true" tabindex="-1"></a>            plt.savefig(output_dir <span class="op">/</span> <span class="ss">f"</span><span class="sc">{</span>tf<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>seqname<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>start<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>end<span class="sc">}</span><span class="ss">_entireSeq_DeepLift.pdf"</span>)</span>
<span id="cb31-183"><a href="#cb31-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-184"><a href="#cb31-184" aria-hidden="true" tabindex="-1"></a>            <span class="co"># zoomed into motif region</span></span>
<span id="cb31-185"><a href="#cb31-185" aria-hidden="true" tabindex="-1"></a>            plot_weights(contr[:, :,dist_start : (dist_start <span class="op">+</span> width<span class="op">+</span><span class="dv">1</span>)],</span>
<span id="cb31-186"><a href="#cb31-186" aria-hidden="true" tabindex="-1"></a>            fontsizes<span class="op">=</span>[<span class="dv">20</span>,<span class="dv">15</span>,<span class="dv">15</span>],</span>
<span id="cb31-187"><a href="#cb31-187" aria-hidden="true" tabindex="-1"></a>            title <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>tf<span class="sc">}</span><span class="ss"> - Motif of interest"</span>, </span>
<span id="cb31-188"><a href="#cb31-188" aria-hidden="true" tabindex="-1"></a>            xlabel<span class="op">=</span><span class="ss">f"</span><span class="sc">{</span>seqname<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span>start<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>end<span class="sc">}</span><span class="ss">, (</span><span class="sc">{</span>dist_start<span class="sc">}</span><span class="ss"> - </span><span class="sc">{</span>dist_start <span class="op">+</span> width<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">)"</span>, </span>
<span id="cb31-189"><a href="#cb31-189" aria-hidden="true" tabindex="-1"></a>            ylabel<span class="op">=</span><span class="st">"DeepLift contribution scores"</span>,</span>
<span id="cb31-190"><a href="#cb31-190" aria-hidden="true" tabindex="-1"></a>            subticks_frequency<span class="op">=</span><span class="dv">10</span>, figsize<span class="op">=</span>figsize2)</span>
<span id="cb31-191"><a href="#cb31-191" aria-hidden="true" tabindex="-1"></a>            plt.savefig(output_dir <span class="op">/</span> <span class="ss">f"</span><span class="sc">{</span>tf<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>seqname<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>start<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>end<span class="sc">}</span><span class="ss">_zoomedSeq_DeepLift.pdf"</span>)  </span>
<span id="cb31-192"><a href="#cb31-192" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb31-193"><a href="#cb31-193" aria-hidden="true" tabindex="-1"></a>            <span class="co"># plot profiles</span></span>
<span id="cb31-194"><a href="#cb31-194" aria-hidden="true" tabindex="-1"></a>            fig, axis <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">2</span>,figsize<span class="op">=</span>(<span class="dv">12</span>,<span class="dv">4</span>))</span>
<span id="cb31-195"><a href="#cb31-195" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].plot(tf_counts[tf_index, <span class="dv">0</span>, :], label<span class="op">=</span><span class="st">"true counts"</span>, color<span class="op">=</span><span class="st">"green"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb31-196"><a href="#cb31-196" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].plot(<span class="op">-</span>tf_counts[tf_index, <span class="dv">1</span>, :], color<span class="op">=</span><span class="st">"green"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb31-197"><a href="#cb31-197" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].plot(pred[tf_index, <span class="dv">0</span>, :], label<span class="op">=</span><span class="st">"pred"</span>, color<span class="op">=</span><span class="st">"blue"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb31-198"><a href="#cb31-198" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].plot(<span class="op">-</span>pred[tf_index, <span class="dv">1</span>, :], color<span class="op">=</span><span class="st">"blue"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)   </span>
<span id="cb31-199"><a href="#cb31-199" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].set_xlabel(<span class="st">"bp"</span>)</span>
<span id="cb31-200"><a href="#cb31-200" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].set_ylabel(<span class="st">"Read counts"</span>)</span>
<span id="cb31-201"><a href="#cb31-201" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].plot(pred[tf_index, <span class="dv">0</span>, :], label<span class="op">=</span><span class="st">"pred"</span>, color<span class="op">=</span><span class="st">"blue"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb31-202"><a href="#cb31-202" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].plot(<span class="op">-</span>pred[tf_index, <span class="dv">1</span>, :], color<span class="op">=</span><span class="st">"blue"</span>, linewidth<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb31-203"><a href="#cb31-203" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].set_xlabel(<span class="st">"bp"</span>)</span>
<span id="cb31-204"><a href="#cb31-204" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].set_ylabel(<span class="st">"Predicted probabilitiy * total counts"</span>)</span>
<span id="cb31-205"><a href="#cb31-205" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">0</span>].legend()</span>
<span id="cb31-206"><a href="#cb31-206" aria-hidden="true" tabindex="-1"></a>            axis[<span class="dv">1</span>].legend()</span>
<span id="cb31-207"><a href="#cb31-207" aria-hidden="true" tabindex="-1"></a>            plt.show()</span>
<span id="cb31-208"><a href="#cb31-208" aria-hidden="true" tabindex="-1"></a>    plot_df <span class="op">=</span> pd.concat(df_list)</span>
<span id="cb31-209"><a href="#cb31-209" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb31-210"><a href="#cb31-210" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> contr, dist_start, plot_df</span>
<span id="cb31-211"><a href="#cb31-211" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-212"><a href="#cb31-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-213"><a href="#cb31-213" aria-hidden="true" tabindex="-1"></a><span class="fu">### Klf4 E2 enhancer (Supplementary Fig.2)</span></span>
<span id="cb31-214"><a href="#cb31-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-217"><a href="#cb31-217" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-218"><a href="#cb31-218" aria-hidden="true" tabindex="-1"></a>contr, dist_start, pred <span class="op">=</span> get_contr_region(<span class="st">"chr4"</span>, start<span class="op">=</span><span class="dv">55475545</span>, end<span class="op">=</span><span class="dv">55475604</span>, dataset<span class="op">=</span>tune_dataset, output_dir<span class="op">=</span>FIG_DIR, dl<span class="op">=</span>dl, tf_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>], device<span class="op">=</span>device,   model<span class="op">=</span>model)</span>
<span id="cb31-219"><a href="#cb31-219" aria-hidden="true" tabindex="-1"></a><span class="co">#pred.to_csv("/home/kathi/AML_Project/data/test_fig.csv")</span></span>
<span id="cb31-220"><a href="#cb31-220" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-221"><a href="#cb31-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-222"><a href="#cb31-222" aria-hidden="true" tabindex="-1"></a><span class="fu">### Nanog enhancer (Supplementary Fig.2)</span></span>
<span id="cb31-223"><a href="#cb31-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-226"><a href="#cb31-226" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-227"><a href="#cb31-227" aria-hidden="true" tabindex="-1"></a>contr, dist_start, pred <span class="op">=</span> get_contr_region(<span class="st">"chr6"</span>, start<span class="op">=</span><span class="dv">122707394</span>, end<span class="op">=</span><span class="dv">122707454</span>, dataset<span class="op">=</span>train_dataset, output_dir<span class="op">=</span>FIG_DIR, dl<span class="op">=</span>dl, tf_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>], device<span class="op">=</span>device, model<span class="op">=</span>model)</span>
<span id="cb31-228"><a href="#cb31-228" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-229"><a href="#cb31-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-230"><a href="#cb31-230" aria-hidden="true" tabindex="-1"></a><span class="fu">### Fbx15 enhancer (Supplementary Fig.2)</span></span>
<span id="cb31-231"><a href="#cb31-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-234"><a href="#cb31-234" aria-hidden="true" tabindex="-1"></a><span class="in">```{python}</span></span>
<span id="cb31-235"><a href="#cb31-235" aria-hidden="true" tabindex="-1"></a>contr, dist_start, pred <span class="op">=</span> get_contr_region(<span class="st">"chr18"</span>, start<span class="op">=</span><span class="dv">84934461</span>, end<span class="op">=</span><span class="dv">84934521</span>, dataset<span class="op">=</span>train_dataset, output_dir<span class="op">=</span>FIG_DIR, dl<span class="op">=</span>dl, tf_list<span class="op">=</span>conf_dict[<span class="st">"tf_list"</span>], device<span class="op">=</span>device, model<span class="op">=</span>model)</span>
<span id="cb31-236"><a href="#cb31-236" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-237"><a href="#cb31-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-238"><a href="#cb31-238" aria-hidden="true" tabindex="-1"></a><span class="fu"># Input x Gradient</span></span>
<span id="cb31-239"><a href="#cb31-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-240"><a href="#cb31-240" aria-hidden="true" tabindex="-1"></a>For the same exemplary regions from the paper.</span>
<span id="cb31-241"><a href="#cb31-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-242"><a href="#cb31-242" aria-hidden="true" tabindex="-1"></a><span class="in">```#{python}</span></span>
<span id="cb31-243"><a href="#cb31-243" aria-hidden="true" tabindex="-1"></a><span class="in">grad, grad_in = input_gradient("chr4", start=55475545, end=55475604, dataset=tune_dataset,output_dir=FIG_DIR, </span></span>
<span id="cb31-244"><a href="#cb31-244" aria-hidden="true" tabindex="-1"></a><span class="in">model=model, tf_list=conf_dict["tf_list"], device=device)</span></span>
<span id="cb31-245"><a href="#cb31-245" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-246"><a href="#cb31-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-247"><a href="#cb31-247" aria-hidden="true" tabindex="-1"></a><span class="fu">### Klf4 E2 enhancer (Supplementary Fig.2)</span></span>
<span id="cb31-248"><a href="#cb31-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-249"><a href="#cb31-249" aria-hidden="true" tabindex="-1"></a><span class="in">```#{python}</span></span>
<span id="cb31-250"><a href="#cb31-250" aria-hidden="true" tabindex="-1"></a><span class="in">grad, grad_in = input_gradient("chr4", start=55475545, end=55475604, dataset=tune_dataset,output_dir=FIG_DIR, </span></span>
<span id="cb31-251"><a href="#cb31-251" aria-hidden="true" tabindex="-1"></a><span class="in"> model=model, tf_list=conf_dict["tf_list"], device=device)</span></span>
<span id="cb31-252"><a href="#cb31-252" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-253"><a href="#cb31-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-254"><a href="#cb31-254" aria-hidden="true" tabindex="-1"></a><span class="fu">### Nanog enhancer (Supplementary Fig.2)</span></span>
<span id="cb31-255"><a href="#cb31-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-256"><a href="#cb31-256" aria-hidden="true" tabindex="-1"></a><span class="in">```#{python}</span></span>
<span id="cb31-257"><a href="#cb31-257" aria-hidden="true" tabindex="-1"></a><span class="in">grad, grad_in = input_gradient("chr6", start=122707394, end=122707454, dataset=train_dataset, output_dir=FIG_DIR, </span></span>
<span id="cb31-258"><a href="#cb31-258" aria-hidden="true" tabindex="-1"></a><span class="in">model=model, tf_list=conf_dict["tf_list"], device=device)</span></span>
<span id="cb31-259"><a href="#cb31-259" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-260"><a href="#cb31-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-261"><a href="#cb31-261" aria-hidden="true" tabindex="-1"></a><span class="fu">### Fbx15 enhancer (Supplementary Fig.2)</span></span>
<span id="cb31-262"><a href="#cb31-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-263"><a href="#cb31-263" aria-hidden="true" tabindex="-1"></a><span class="in">```#{python}</span></span>
<span id="cb31-264"><a href="#cb31-264" aria-hidden="true" tabindex="-1"></a><span class="in">grad, grad_in = input_gradient("chr17", start=35504453, end=35504603, dataset=train_dataset, output_dir=FIG_DIR, </span></span>
<span id="cb31-265"><a href="#cb31-265" aria-hidden="true" tabindex="-1"></a><span class="in">model=model, tf_list=conf_dict["tf_list"], device=device)</span></span>
<span id="cb31-266"><a href="#cb31-266" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




</body></html>